{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(reshape2)\n",
    "library(RColorBrewer)\n",
    "suppressMessages(library(dplyr))\n",
    "library(stringr)\n",
    "suppressMessages(library(tidyr))\n",
    "theme_set(theme_bw())\n",
    "library(scales)\n",
    "options(repr.plot.width=7, repr.plot.height=4)\n",
    "isotypes = c('Ala', 'Arg', 'Asn', 'Asp', 'Cys', 'Gln', 'Glu', 'Gly', 'His', 'Ile', 'iMet', 'Leu', 'Lys', 'Met', 'Phe', 'Pro', 'Ser', 'Thr', 'Trp', 'Tyr', 'Val')\n",
    "\n",
    "fills = c('A'='#ffd92f', 'C'='#4daf4a', 'G'='#e41a1c', 'U'='#377eb8', 'A:U'='#93da69', 'U:A'='#93da69', 'G:C'='#c1764a', 'C:G'='#c1764a', 'G:U'='#b26cbd', 'U:G'='#b26cbd', '-'='gray60', '-:-'='gray60')\n",
    "suppressMessages(library(Biostrings))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "load('best-freqs.RData')\n",
    "load('clade-isotype-specific.RData')\n",
    "load('isotype-specific.RData')\n",
    "load('consensus-IDEs.RData')\n",
    "load('clade-isotype-specific-freqs.RData')\n",
    "identities = read.delim('identities.tsv', sep='\\t')\n",
    "identities$quality = as.logical(identities$quality)\n",
    "identities$restrict = as.logical(identities$restrict)\n",
    "identities = identities %>% mutate(quality=quality & !restrict)\n",
    "genome_table = read.delim('genome_table+.txt', sep='\\t', stringsAsFactors=FALSE, header=FALSE, col.names=c(\"species_short\", \"species\", \"species_long\", \"domain\", \"clade\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Weird tRNAs in context of sequence features\n",
    "\n",
    "List of weird tRNAs:\n",
    "\n",
    "- Susan Ackerman's [Arg-UCU](http://gtrnadb.ucsc.edu/genomes/eukaryota/Hsapi19/genes/tRNA-Arg-TCT-4-1.html) in mouse. \n",
    "    - Only Arg-UCU without an intron.\n",
    "    - See alignment for details, but it also includes C20a instead of T20a, G20b, C40 instead of T40, etc. \n",
    "    - Most importantly, it contains C50, but its mutation to T50 likely causes misfolding in the B6J mouse strain.\n",
    "    - Conserved in human, and potentially all the way to zebrafish and Drosophila.\n",
    "- [Lys-CUU](http://gtrnadb.ucsc.edu/genomes/eukaryota/Scere3/genes/tRNA-Lys-CTT-1-1.html) in yeast. It's the only isodecoder for Lys-CUU. Can fold into an \"F\"-form able to be imported into the mitochondria.\n",
    "- [Gly-GCC](http://gtrnadb.ucsc.edu/genomes/eukaryota/Hsapi19/genes/tRNA-Gly-GCC-2-1.html), which is downregulated in B cell lymphoma cell lines (Maute paper). Represses RPA1, which is involved in DNA repair, DNA replication, and DNA damage response. Overexpression reduces proliferation and sensitizes cell to etoposide-induced DNA damage.\n",
    "\n",
    "## Code for extracting position-specific scores\n",
    "\n",
    "We need to (a) filter for the subset of tRNAs we're interested in, (b) create a multiple alignment -> covariance model (no need to calibrate), (c) align, (d) parse output, and (e) visualize."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "calculate_position_specific_scores = function(seq = \"\", seqname = \"\", clade = \"\", isotype = \"\", anticodon = \"\") {\n",
    "  # get subset of tRNAs and write to file\n",
    "  subset = identities %>% select_('species', 'seqname', 'isotype', 'clade', 'anticodon', 'quality') %>% filter_('quality')\n",
    "  if (clade != \"\") subset = subset %>% filter_(paste0(\"clade == '\", clade, \"'\"))\n",
    "  if (isotype != \"\") subset = subset %>% filter_(paste0(\"isotype == '\", isotype, \"'\"))\n",
    "  if (isotype != \"\" & anticodon != \"\") subset = subset %>% filter_(paste0(\"anticodon == '\", anticodon, \"'\"))\n",
    "  euk_seqs = readDNAStringSet(filepath = '/projects/lowelab/users/blin/identity/euk-isotypes/fasta/euk.fa', format = 'fasta')\n",
    "  writeXStringSet(euk_seqs[match(subset$seqname, str_extract(names(euk_seqs), '\\\\S+'))],\n",
    "                  filepath = 'subset.fa')\n",
    "\n",
    "  # create covariance model\n",
    "  system('cmalign -g --notrunc --matchonly -o subset.sto /projects/lowelab/users/blin/tRNAscan/models/domain-specific/euk-num-092016.cm subset.fa')\n",
    "  system('cmbuild --hand --enone -F subset.cm subset.sto')\n",
    "\n",
    "  # align our tRNA\n",
    "  seq = DNAStringSet(seq)\n",
    "  names(seq) = seqname\n",
    "  writeXStringSet(seq, filepath = \"my-tRNA.fa\")\n",
    "  system('cmalign -g --notrunc --matchonly --tfile my-tRNA.tfile -o my-tRNA.sto subset.cm my-tRNA.fa')\n",
    "    \n",
    "  # parse output\n",
    "  system('python parse-parsetree.py my-tRNA.tfile > my-tRNA.bits')\n",
    "  bits = read.table('my-tRNA.bits', header = FALSE) %>%\n",
    "    mutate(Position = factor(V1, c('1:72', '2:71', '3:70', '4:69', '5:68', '6:67', '7:66', '8', '9', '10:25', '11:24', '12:23', '13:22', '14', '15', '16', '17', '18', '19', '20', '20a', '21', '26', '27:43', '28:42', '29:41', '30:40', '31:39', '32', '33', '34', '35', '36', '37', '38', '44', '45', '46', '47', '48', '49:65', '50:64', '51:63', '52:62', '53:61', '54', '55', '56', '57', '58', '59', '60', '73'))) %>%\n",
    "    mutate(Bits = V2) %>%\n",
    "    mutate(Identity = V3) %>%\n",
    "    mutate(Source = \"Input\") %>%\n",
    "    select(-V1, -V2) %>%\n",
    "    filter(!is.na(Position))\n",
    "\n",
    "  # emit and read in consensus sequence\n",
    "  # cmemit is not good at putting deletions/insertions into consensus emit (-c switch), so exponentiate instead\n",
    "  system('cmemit -N 1 --exp 7 -o subset-cons.fa subset.cm')\n",
    "  system('cmalign -g --notrunc --matchonly --tfile subset-cons.tfile -o subset-cons.sto subset.cm subset-cons.fa')\n",
    "  system('python parse-parsetree.py subset-cons.tfile > subset-cons.bits')\n",
    "\n",
    "  bits = rbind(bits, read.table('subset-cons.bits', header = FALSE) %>%\n",
    "    mutate(Position = factor(V1, c('1:72', '2:71', '3:70', '4:69', '5:68', '6:67', '7:66', '8', '9', '10:25', '11:24', '12:23', '13:22', '14', '15', '16', '17', '18', '19', '20', '20a', '21', '26', '27:43', '28:42', '29:41', '30:40', '31:39', '32', '33', '34', '35', '36', '37', '38', '44', '45', '46', '47', '48', '49:65', '50:64', '51:63', '52:62', '53:61', '54', '55', '56', '57', '58', '59', '60', '73'))) %>%\n",
    "    mutate(Bits = V2) %>%\n",
    "    mutate(Identity = V3) %>%\n",
    "    mutate(Source = \"Consensus\") %>%\n",
    "    select(-V1, -V2) %>%\n",
    "    filter(!is.na(Position)))\n",
    "    \n",
    "  # compare consensus and our tRNA; fix instances where cmemit did not output most likely identity\n",
    "  bits = bits %>% group_by(Position) %>%\n",
    "    arrange(Source) %>% \n",
    "    summarize(Bits = Bits[2] - Bits[1], Consensus = Identity[1], Identity = Identity[2]) %>% \n",
    "    ungroup() %>%\n",
    "    mutate(Consensus = ifelse(Bits > 0, as.character(Identity), as.character(Consensus))) %>%\n",
    "    mutate(Bits = ifelse(Bits > 0, 0, Bits))\n",
    "\n",
    "  # clean up\n",
    "  system('rm subset.sto subset.cm subset.fa')\n",
    "  system('rm subset-cons.fa subset-cons.bits subset-cons.tfile subset-cons.sto')\n",
    "  system('rm my-tRNA.fa my-tRNA.tfile my-tRNA.bits my-tRNA.sto')\n",
    "\n",
    "  return(bits)\n",
    "}\n",
    "    \n",
    "calculate_scores_multiplex = function(seq = \"\", seqname = \"\", clade = \"\", isotype = \"\", anticodon = \"\") {\n",
    "\n",
    "  bits = calculate_position_specific_scores(seq, seqname, clade = \"\", isotype = \"\", anticodon = \"\") %>% \n",
    "    mutate(Clade = \"Eukaryota\", Isotype = \"All isotypes\", Anticodon = \"\")\n",
    "\n",
    "  if (clade != \"\") {\n",
    "    bits = rbind(bits, calculate_position_specific_scores(seq, seqname, clade, isotype = \"\", anticodon = \"\") %>%\n",
    "                 mutate(Clade = clade, Isotype = \"All isotypes\", Anticodon = \"\"))\n",
    "  }\n",
    "  if (isotype != \"\") {\n",
    "    bits = rbind(bits, calculate_position_specific_scores(seq, seqname, clade = \"\", isotype, anticodon = \"\") %>%\n",
    "                 mutate(Clade = \"Eukaryota\", Isotype = isotype, Anticodon = \"(all isodecoders)\"))\n",
    "  }\n",
    "  if (clade != \"\" & isotype != \"\") {\n",
    "    bits = rbind(bits, calculate_position_specific_scores(seq, seqname, clade = clade, isotype = isotype, anticodon = \"\") %>%\n",
    "                 mutate(Clade = clade, Isotype = isotype, Anticodon = \"(all isodecoders)\"))\n",
    "  }\n",
    "  if (isotype != \"\" & anticodon != \"\") {\n",
    "    bits = rbind(bits, calculate_position_specific_scores(seq, seqname, clade = \"\", isotype = isotype, anticodon = anticodon) %>%\n",
    "                 mutate(Clade = \"Eukaryota\", Isotype = isotype, Anticodon = anticodon))\n",
    "  }\n",
    "  if (clade != \"\" & isotype != \"\" & anticodon != \"\") {\n",
    "    bits = rbind(bits, calculate_position_specific_scores(seq, seqname, clade, isotype, anticodon) %>%\n",
    "                 mutate(Clade = clade, Isotype = isotype, Anticodon = anticodon))\n",
    "  }\n",
    "  \n",
    "  return(bits %>% mutate(tRNA = seqname))\n",
    "}\n",
    "  \n",
    "calculate_scores_multiseq = function(seqs = \"\", clade = \"Eukaryota\", isotype = \"\", anticodon = \"\") {  \n",
    "  multi_bits = data.frame(Position = character(0), Bits = character(0), Identity = character(0), Consensus = character(0), Clade = character(0), Isotype = character(0), Anticodon = character(0))\n",
    "    for (i in 1:length(seqs)) {\n",
    "    bits = calculate_position_specific_scores(seqs[i], names(seqs)[i], clade, isotype, anticodon) %>% \n",
    "      mutate(Clade = clade, Isotype = isotype, Anticodon = anticodon, tRNA = names(seqs)[i])\n",
    "    multi_bits = rbind(multi_bits, bits)\n",
    "  }\n",
    "  \n",
    "  return(multi_bits)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Arg-UCU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=10, repr.plot.height=13)\n",
    "bits = calculate_scores_multiplex(seq = \"GTCTCTGTGGCGCAATGGAcgAGCGCGCTGGACTTCTAATCCAGAGGtTCCGGGTTCGAGTCCCGGCAGAGATG\",\n",
    "                                  seqname = \"hg19_chr1.trna84-ArgTCT\",\n",
    "                                  clade = \"Mammalia\", \n",
    "                                  isotype = \"Arg\",\n",
    "                                  anticodon = \"TCT\")\n",
    "\n",
    "df = bits %>% mutate(Clade = as.factor(Clade), Isotype = as.factor(Isotype), Anticodon = as.factor(Anticodon), Consensus = as.factor(Consensus), Identity = as.factor(Identity), tRNA = as.factor(tRNA)) %>%\n",
    "  mutate(Isodecoder = as.factor(paste0(Isotype, \" \", Anticodon))) %>%\n",
    "  select(Position, Clade, tRNA, Bits, Consensus, Identity, Isodecoder)\n",
    "\n",
    "ggplot(df) + geom_bar(aes(x = Position, y = Bits, fill = Identity), size = 0.1, color = 'gray20', width = 0.8, position = 'dodge', stat = 'identity') + \n",
    "  facet_wrap( ~ paste0(Clade, \" / \", Isodecoder), ncol = 3) + \n",
    "  geom_text(aes(x = Position, color = Identity, label = Identity), size = 3, y = min(df$Bits) - 0.4) +\n",
    "  geom_text(aes(x = Position, color = Consensus, label = Consensus), size = 3, y = max(df$Bits) + 0.4) +\n",
    "  coord_flip() +\n",
    "  scale_fill_manual(values=fills) +\n",
    "  scale_color_manual(values=fills) + \n",
    "  scale_x_discrete(limits = rev(levels(df$Position))) + \n",
    "  scale_y_continuous(limits = c(min(df$Bits) - 0.6, max(df$Bits) + 0.6)) + \n",
    "  theme(legend.position='none') + \n",
    "  ylab(\"Score difference between consensus and input (bits)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=12, repr.plot.height=7)\n",
    "bits = calculate_scores_multiseq(seqs = c(\"hg19-tRNA-Arg-TCT-4-1\" = \"GTCTCTGTGGCGCAATGGAcgAGCGCGCTGGACTTCTAATCCAGAGGtTCCGGGTTCGAGTCCCGGCAGAGATG\",\n",
    "                                          \"hg19-tRNA-Arg-TCT-1-1\" = \"GGCTCCGTGGCGCAATGGAtAGCGCATTGGACTTCTAgaggctgaaggcATTCAAAGGtTCCGGGTTCGAGTCCCGGCGGAGTCG\",\n",
    "                                          \"hg19-tRNA-Arg-TCT-2-1\" = \"GGCTCTGTGGCGCAATGGAtAGCGCATTGGACTTCTAgtgacgaatagagcaATTCAAAGGtTGTGGGTTCGAATCCCACCAGAGTCG\",\n",
    "                                          \"hg19-tRNA-Arg-TCT-3-1\" = \"GGCTCTGTGGCGCAATGGAtAGCGCATTGGACTTCTAgctgagcctagtgtggtcATTCAAAGGtTGTGGGTTCGAGTCCCACCAGAGTCG\"),\n",
    "                                 clade = \"Mammalia\", isotype = \"Arg\", anticodon = \"TCT\")\n",
    "\n",
    "bits %>% select(Position, Bits, Identity, Consensus, tRNA) %>%\n",
    "  ggplot() + geom_bar(aes(x = Position, y = Bits, fill = Identity), size = 0.1, color = 'gray20', width = 0.8, position = 'dodge', stat = 'identity') + \n",
    "  facet_wrap( ~ tRNA, nrow = 1) +\n",
    "  geom_text(aes(x = Position, color = Identity, label = Identity), size = 3, y = min(bits$Bits) - 0.4) +\n",
    "  geom_text(aes(x = Position, color = Consensus, label = Consensus), size = 3, y = max(bits$Bits) + 0.4) +\n",
    "  coord_flip() +\n",
    "  scale_fill_manual(values=fills) +\n",
    "  scale_color_manual(values=fills) + \n",
    "  scale_x_discrete(limits = rev(levels(bits$Position))) + \n",
    "  scale_y_continuous(limits = c(min(bits$Bits) - 0.6, max(bits$Bits) + 0.6)) + \n",
    "  theme(legend.position='none') + \n",
    "  ylab(\"Score difference between consensus and input (bits)\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
