{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from ggplot import *\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "mpl.rcParams['figure.figsize'] = (15, 6)\n",
    "mpl.style.use('ggplot')\n",
    "import pandas as pd\n",
    "import numpy\n",
    "import subprocess\n",
    "import glob\n",
    "import re\n",
    "import os, sys\n",
    "from collections import defaultdict, Counter, OrderedDict\n",
    "from Bio import SeqIO, SeqRecord, Seq\n",
    "from tRNA_position import *\n",
    "pd.set_option('display.max_colwidth',10000)\n",
    "pd.set_option('display.width', 10000)\n",
    "pd.set_option('display.max_columns', 10000)\n",
    "pd.set_option('display.max_rows',1000)\n",
    "isotypes = ['Ala', 'Arg', 'Asn', 'Asp', 'Cys', 'Gln', 'Glu', 'Gly', 'His', 'Ile', 'iMet', 'Leu', 'Lys', 'Met', 'Phe', 'Pro', 'Ser', 'Thr', 'Trp', 'Tyr', 'Val']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "A global view of identity elements versus biological features would be a powerful tool for predicting tRNA function using primary sequence. To do this, I'll need to align eukaryotic tRNAs to our established models and annotate positions based on universal numbering. Then, I'll extract sequence information such as position, clade, or variable arm length."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process tRNAs\n",
    "## Species information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>longname</th>\n",
       "      <th>domain</th>\n",
       "      <th>clade</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ashbGoss_ATCC10895</td>\n",
       "      <td>Eremothecium gossypii ATCC 10895</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>33169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aspeFumi_AF293</td>\n",
       "      <td>Aspergillus fumigatus Af293</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>746128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aspeNidu_FGSC_A4</td>\n",
       "      <td>Aspergillus nidulans FGSC A4</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>162425.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aspeOryz_RIB40</td>\n",
       "      <td>Aspergillus oryzae RIB40</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>5062.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>botrCine_B05_10</td>\n",
       "      <td>Botrytis cinerea B05.10</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>40559.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              species                          longname     domain  clade     taxid\n",
       "0  ashbGoss_ATCC10895  Eremothecium gossypii ATCC 10895  eukaryota  Fungi   33169.0\n",
       "1      aspeFumi_AF293       Aspergillus fumigatus Af293  eukaryota  Fungi  746128.0\n",
       "2    aspeNidu_FGSC_A4      Aspergillus nidulans FGSC A4  eukaryota  Fungi  162425.0\n",
       "3      aspeOryz_RIB40          Aspergillus oryzae RIB40  eukaryota  Fungi    5062.0\n",
       "4     botrCine_B05_10           Botrytis cinerea B05.10  eukaryota  Fungi   40559.0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_table = pd.read_table('genomes-051017.tsv', header=None, names=['species', 'longname', 'domain', 'clade', 'taxid'])\n",
    "species_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align tRNAs to Sprinzl numbering model\n",
    "\n",
    "We have a \"quality set\" of tRNAs, which include tRNAs from previous tRNAscan-SE runs and newer runs. We also want to keep track of high-scoring, isotype mismatched tRNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "species = sorted(glob.glob(\"out/*-detailed.out\"))\n",
    "species = [sp[4:-19] for sp in species]\n",
    "seqs = []\n",
    "for sp in species:  \n",
    "  sys.stdout.write('processing ' + sp + '...')\n",
    "  sys.stdout.flush()\n",
    "  \n",
    "  tRNA_file = 'tRNAs/' + sp + '-tRNAs.fa'\n",
    "  tscanout_file = 'out/{}-tRNAs-detailed.out'.format(sp)\n",
    "  # Import annotation from tRNAscan .out files. We want to keep the ones that are \"Quality set\" or \"Unexpected anticodon;First-pass quality filtered\".\n",
    "  approved_tRNAs = []\n",
    "  for metadata in pd.read_table(tscanout_file, sep=\"\\t\", skiprows=3, header=None).iterrows():\n",
    "    if (metadata[1].iloc[-1] in [\"Quality set\", \"Unexpected anticodon;First-pass quality filtered\", \"quality set\", \"unexpected anticodon,first-pass quality filtered\", \"Isotype mismatch;First-pass quality filtered\"]\n",
    "        or re.match('^IPD:-\\d+\\.\\d+,isotype mismatch,first-pass quality filtered$', metadata[1].iloc[-1])):\n",
    "      approved_tRNAs.append('{}.trna{}-{}{}'.format(metadata[1][0], metadata[1][1], metadata[1][4], metadata[1][5]))\n",
    " \n",
    "  # Remove introns using sstofa\n",
    "  subprocess.call('sstofa3 ss/{}-tRNAs.ss \"\" 1 0 > {}'.format(sp, tRNA_file), shell=True)\n",
    "\n",
    "  # Parse isotype-specific scores file\n",
    "  iso_scores = pd.read_table('iso/' + sp + '-tRNAs.iso', header=0)\n",
    "  iso_scores['best'] = iso_scores.ix[:,2:].idxmax(axis=1)\n",
    "  iso_scores['score'] = iso_scores.max(axis=1, numeric_only=True)\n",
    "  iso_scores.index = iso_scores.tRNAscanID.values\n",
    "  iso_scores = iso_scores[['best', 'score']]\n",
    "  \n",
    "  # Parse seqs, filtering out the ones we don't need, and add to proper isotype\n",
    "  for seq in SeqIO.parse('tRNAs/{}-tRNAs.fa'.format(sp), 'fasta'):\n",
    "    if seq.id not in approved_tRNAs: continue\n",
    "    if \"pseudogene\" in seq.description: continue\n",
    "    score = float(re.findall('Sc: [\\d\\.]+', seq.description)[0].split()[-1])\n",
    "    if score < 55: continue\n",
    "    trnascanid = re.findall('.+\\.trna\\d+', seq.id)[0]\n",
    "    isotype = iso_scores.ix[trnascanid].best\n",
    "    if isotype == 'SeC' or 'mito' in isotype or 'mt' in trnascanid: continue\n",
    "    isoscore = iso_scores.ix[trnascanid].score\n",
    "    seq.id = '{}|{} Iso: {} ({})'.format(sp, seq.description, isoscore, isotype)\n",
    "    seq.description = ''\n",
    "    seqs.append(seq)\n",
    "    \n",
    "  print('finished')\n",
    "  sys.stdout.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fasta_handle = open('euk-tRNAs.fa', 'w')\n",
    "SeqIO.write(seqs, fasta_handle, 'fasta')\n",
    "fasta_handle.close()\n",
    "num_model = '/projects/lowelab/users/blin/tRNAscan/models/domain-specific/euk-num-092016.cm'\n",
    "subprocess.call('cmalign -g --notrunc -o euk-tRNAs.sto {} euk-tRNAs.fa'.format(num_model), shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create table of tRNA bases by position\n",
    "\n",
    "This is a giant data frame with one row per tRNA, and with columns for each position, plus tRNA metadata like species and loop lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skipped the following: {'saccCere_VIN13', 'saccCere_W303', 'saccCere_UFMG_A_905', 'saccCere_R008', 'saccCere_P283', 'saccCere_FOSTERSB', 'saccCere_P301', 'saccCere_VL3', 'saccCere_YJSH1', 'saccCere_CEN_PK113_7D', 'saccCere_R103', 'saccCere_SIGMA1278B', 'saccCere_AWRI796'}\n"
     ]
    }
   ],
   "source": [
    "def position_base(positions, seq):\n",
    "  for position_index, position in enumerate(positions):\n",
    "    if position.paired:\n",
    "      index1, index2 = position.position.split(':')\n",
    "      index1, index2 = int(index1), int(index2)\n",
    "      base_pair = \"{}:{}\".format(seq[index1 - 1].upper(), seq[index2 - 1].upper())\n",
    "      yield position.sprinzl, base_pair\n",
    "    else:\n",
    "      index = int(position.position)\n",
    "      base = seq[index - 1].upper()\n",
    "      yield position.sprinzl, base\n",
    "      \n",
    "identities = pd.DataFrame()\n",
    "\n",
    "# get positions\n",
    "alignment_fhandle = open('euk-tRNAs.sto')\n",
    "positions = [] # list containing each position in the tRNA\n",
    "\n",
    "# first, get secondary structure\n",
    "for line in alignment_fhandle:\n",
    "  if line[0:12] == '#=GC SS_cons':\n",
    "    ss = line.strip().split()[-1]\n",
    "alignment_fhandle.close()\n",
    "\n",
    "# parse secondary structure into regions and positions\n",
    "positions = annotate_positions(ss)\n",
    "\n",
    "# get nucleotide at each position for each tRNA by parsing Stockholm file\n",
    "alignment_fhandle = open('euk-tRNAs.sto')\n",
    "trnas = []\n",
    "skipped = []\n",
    "for line in alignment_fhandle:\n",
    "  if line[0] in [\"#\", '\\n', '/']: continue\n",
    "  species, desc = line.strip().split('|', 1)\n",
    "  seqname = desc.split()[0]\n",
    "  seq = desc.split()[-1]\n",
    "  if any(species_table.species == species):\n",
    "    row = species_table[species_table.species == species]\n",
    "  else: \n",
    "    skipped.append(species)\n",
    "    continue\n",
    "  isotype = re.findall('\\.trna\\d+-([A-Za-z]+)', seqname)[0][:-3]\n",
    "  seqname = '{}_{}'.format(species, seqname)\n",
    "  trna = {'domain': row.domain.values[0], 'clade': row.clade.values[0], 'species': species, 'species_long': row.longname.values[0], 'taxid': row.taxid.values[0], 'seqname': seqname, 'isotype': isotype}\n",
    "  trna = {**trna, **{sprinzl: base for sprinzl, base in position_base(positions, seq)}}\n",
    "  trnas.append(trna)\n",
    "\n",
    "if len(skipped) > 0: print('skipped the following: {}'.format(set(skipped)))\n",
    "identities = identities.append(trnas, ignore_index=True)\n",
    "identities.fillna('.', inplace=True)\n",
    "alignment_fhandle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create single base columns from paired positions\n",
    "\n",
    "This is mainly for plotting and such."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cols = list(filter(lambda x: ':' in x, identities.columns))\n",
    "for col in cols:\n",
    "  pos1, pos2 = col.split(':')\n",
    "  base1 = [bases.split(':')[0] for bases in identities[col]]\n",
    "  base2 = [bases.split(':')[1] for bases in identities[col]]\n",
    "  identities[pos1] = base1\n",
    "  identities[pos2] = base2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Shorten D stem for class II tRNAs\n",
    "\n",
    "Leu and Tyr have a 3 base pair D stem instead of 4. Since we've already copied over the columns into single bases, this is okay. I don't want to change the entries for this position to \"-:-\" because it would imply a deletion, and I don't want to change it to \".:.\" because it would break a lot of code. I'll leave it alone for now, until it's necessary to change it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate tertiary interactions\n",
    "\n",
    "We also want to annotate the 2-base version of the base triples. This makes checking for conserved interactions (e.g. an identity element) a bit easier. It's important to remember to consider the 3rd base pair too, though."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "identities['8:14:21'] = identities.ix[:, [\"8\", \"14\", \"21\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['8:14'] = identities.ix[:, [\"8\", \"14\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['9:12:23'] = identities.ix[:, [\"9\", \"12\", \"23\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['9:23'] = identities.ix[:, [\"9\", \"23\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['10:25:45'] = identities.ix[:, [\"10\", \"25\", \"45\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['10:45'] = identities.ix[:, [\"10\", \"45\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['13:22:46'] = identities.ix[:, [\"13\", \"22\", \"46\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['22:46'] = identities.ix[:, [\"22\", \"46\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['15:48'] = identities.ix[:, [\"15\", \"48\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['18:55'] = identities.ix[:, [\"18\", \"55\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['19:56'] = identities.ix[:, [\"19\", \"56\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['26:44'] = identities.ix[:, [\"26\", \"44\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['54:58'] = identities.ix[:, [\"54\", \"58\"]].apply(lambda row: ':'.join(row), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional sequence information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Isotype, anticodon, score\n",
    "seqinfo = []\n",
    "for line in open('euk-tRNAs.sto'):\n",
    "  if line[0:4] != \"#=GS\": continue\n",
    "  _, desc, _, _, isotype, anticodon, _, _, _, score, _, isoscore, isotype_best = line.strip().split()\n",
    "  seqname = desc.replace('|', '_', 1)\n",
    "  seqinfo.append([seqname, isotype_best[1:-1], anticodon[1:-1], float(score), float(isoscore)])\n",
    "seqinfo = pd.DataFrame(seqinfo, columns=['seqname', 'isotype_best', 'anticodon', 'score', 'isoscore'])\n",
    "identities = identities.merge(seqinfo, on='seqname')\n",
    "\n",
    "# iMet may or may not be properly annotated, depending on tRNAscan-SE run\n",
    "unlifted_imets = identities.ix[(identities.isotype == \"Met\") & (identities.isotype_best == \"iMet\"), ].index\n",
    "identities.ix[unlifted_imets, 'isotype'] = 'iMet'\n",
    "\n",
    "# GC content\n",
    "paired_cols = identities.columns[list(map(lambda x: (':' in x), identities.columns))]\n",
    "identities['GC'] = identities[paired_cols].apply(lambda x: sum((x == \"G:C\") | (x == \"C:G\"))/len(paired_cols), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insertions/deletions/introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Intron length\n",
    "intron_cols = list(filter(lambda x: x[0:3] == '37i', identities.columns))\n",
    "identities['intron'] = identities[intron_cols].apply(lambda x: sum(x != \".\"), axis=1)\n",
    "\n",
    "# Insertions (minus introns at 37/38)\n",
    "insertion_cols = list(filter(lambda x: bool(re.search('^\\d+i', x)) & (x not in intron_cols), identities.columns))\n",
    "identities['insertions'] = identities[insertion_cols].apply(lambda x: sum(x != '.'), axis=1)\n",
    "\n",
    "# Deletions at positions that are not the variable arm, and not counting 17/17a/20a/20b\n",
    "base_cols = list(filter(lambda x: bool(re.match('^\\d+$', x)) & (x not in ['74', '75', '76', '17', '17a', '20a', '20b']), identities.columns))\n",
    "identities['deletions'] = identities[base_cols].apply(lambda x: ''.join(x).count('-'), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loop sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def bounds_to_cols(cols, start, end):\n",
    "  selected_cols = []\n",
    "  for col in cols:\n",
    "    matches = re.findall('\\d+', col)\n",
    "    if len(matches) < 1: continue\n",
    "    index = int(matches[0])\n",
    "    if (index >= start and index <= end or col[0:3] == '{}i'.format(start - 1)) and col[0] != 'V':\n",
    "      selected_cols.append(col)\n",
    "  return selected_cols\n",
    "\n",
    "dloop_cols = list(filter(lambda col: ':' not in col, bounds_to_cols(identities.columns, 14, 21)))\n",
    "identities['D-loop'] = identities[dloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "# Leu, Ser have a 3 bp D stem\n",
    "dloop_II_cols = list(filter(lambda col: ':' not in col, bounds_to_cols(identities.columns, 13, 22)))\n",
    "identities.ix[(identities.isotype == 'Leu') | (identities.isotype == 'Ser'), 'D-loop'] = identities[dloop_II_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "acloop_cols = list(filter(lambda x: not re.match('37i.+', x), bounds_to_cols(identities.columns, 32, 38)))\n",
    "identities['AC-loop'] = identities[acloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "tpcloop_cols = bounds_to_cols(identities.columns, 54, 60)\n",
    "identities['TPC-loop'] = identities[tpcloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "varm_cols = list(filter(lambda x: ('V' in x) & (':' not in x), identities.columns))\n",
    "identities['V-arm'] = identities[varm_cols].apply(lambda x: len(x[(x != '.') & (x != '-') & (x != \"-:-\")]), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### High-quality tRNA set\n",
    "\n",
    "There was talk about using the high quality tRNAs to look at identity elements. Instead, I will just annotate specific tRNAs as high-quality or not. We can sort out the differences later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Import list of quality tRNAs\n",
    "quality_trnas = [line.strip() for line in open('quality-set.out')]\n",
    "identities['quality'] = identities.seqname.isin(quality_trnas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Restrict tRNAs by species\n",
    "\n",
    "We may also want to limit the contribution of any single species, similar to how we built the isotype-specific models. The restricted set is a subset of the quality set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "identities.ix[:, 'restrict'] = False\n",
    "for species in identities.species.unique():\n",
    "  species_df = identities.ix[identities.species == species, :]\n",
    "  for isotype in isotypes:\n",
    "    isotype_indices = identities.ix[(identities.species == species) & (identities.isotype == isotype) & (-identities.restrict), ].index\n",
    "    quality_indices = identities.ix[(identities.species == species) & (identities.isotype == isotype) & (-identities.restrict) & (identities.quality), ].index\n",
    "    unique_scores_indices = isotype_indices[-identities.ix[isotype_indices, 'score'].duplicated()]\n",
    "    unique_quality_indices = quality_indices[-identities.ix[quality_indices, 'score'].duplicated()]\n",
    "    \n",
    "    if len(unique_quality_indices) >= 50:\n",
    "      unique_quality_indices = unique_quality_indices[numpy.argsort(identities.ix[unique_quality_indices, 'score'])][::-1][:50]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_quality_indices))\n",
    "    elif len(unique_quality_indices) > 0:\n",
    "      if len(unique_scores_indices) > 50 - len(unique_quality_indices):\n",
    "        unique_scores_indices = unique_scores_indices[numpy.argsort(identities.ix[unique_scores_indices, 'score'])][::-1][:50 - len(unique_quality_indices)]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_quality_indices) - set(unique_scores_indices))\n",
    "    else:\n",
    "      if len(unique_scores_indices) > 50:\n",
    "        unique_scores_indices = unique_scores_indices[numpy.argsort(identities.ix[unique_scores_indices, 'score'])][::-1][:50]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_scores_indices))\n",
    "    \n",
    "    identities.ix[restricted_indices, 'restrict'] = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The quality set does have some issues, though: it filters out rare codons, when rare is relative to model species. It also performs a hard check for anticodon-isotype model match. Fungi have much more diverged tRNAs and, in many cases, a different isotype scores higher even when it's clear that it's the bona fide tRNA (mostly based on number of tRNAs of that isotype, or codon usage). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export to R\n",
    "\n",
    "R has superior visualization capabilities.\n",
    "\n",
    "### Order columns\n",
    "\n",
    "To make it look pretty."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def position_str_to_int(position):\n",
    "  if position == \"20a\": return 20.1\n",
    "  if position == \"20b\": return 20.2\n",
    "  digits = re.findall('\\d+', position)\n",
    "  if len(digits) == 0: return -1\n",
    "  insert = 0\n",
    "  if 'i' in position and len(digits) == 2: insert = float(digits[1]) / 1000\n",
    "  if position[0] == 'V':\n",
    "    if ':' in position: return int(digits[0]) + 45 - 10 + insert # V11~V17\n",
    "    else: return int(digits[0]) + 45 + 7 + insert # V1~V5\n",
    "  if int(digits[0]) >= 46: return int(digits[0]) + 50 + insert # just add an arbitrarily large number to skip v-arm\n",
    "  return int(digits[0]) + insert\n",
    "\n",
    "identities = identities[sorted(list(identities.columns), key=position_str_to_int)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "identities.to_csv(path_or_buf='identities.tsv', sep='\\t', index_label=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>domain</th>\n",
       "      <th>isotype</th>\n",
       "      <th>seqname</th>\n",
       "      <th>species</th>\n",
       "      <th>species_long</th>\n",
       "      <th>taxid</th>\n",
       "      <th>isotype_best</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>score</th>\n",
       "      <th>isoscore</th>\n",
       "      <th>GC</th>\n",
       "      <th>intron</th>\n",
       "      <th>insertions</th>\n",
       "      <th>deletions</th>\n",
       "      <th>D-loop</th>\n",
       "      <th>AC-loop</th>\n",
       "      <th>TPC-loop</th>\n",
       "      <th>V-arm</th>\n",
       "      <th>quality</th>\n",
       "      <th>restrict</th>\n",
       "      <th>1:72</th>\n",
       "      <th>1</th>\n",
       "      <th>1i1</th>\n",
       "      <th>2:71</th>\n",
       "      <th>2</th>\n",
       "      <th>2i1</th>\n",
       "      <th>3:70</th>\n",
       "      <th>3</th>\n",
       "      <th>3i1</th>\n",
       "      <th>3i2</th>\n",
       "      <th>3i3</th>\n",
       "      <th>3i4</th>\n",
       "      <th>3i5</th>\n",
       "      <th>3i6</th>\n",
       "      <th>4:69</th>\n",
       "      <th>4</th>\n",
       "      <th>4i1</th>\n",
       "      <th>4i2</th>\n",
       "      <th>4i3</th>\n",
       "      <th>4i4</th>\n",
       "      <th>4i5</th>\n",
       "      <th>4i6</th>\n",
       "      <th>4i7</th>\n",
       "      <th>4i8</th>\n",
       "      <th>4i9</th>\n",
       "      <th>4i10</th>\n",
       "      <th>4i11</th>\n",
       "      <th>5:68</th>\n",
       "      <th>5</th>\n",
       "      <th>5i1</th>\n",
       "      <th>5i2</th>\n",
       "      <th>5i3</th>\n",
       "      <th>5i4</th>\n",
       "      <th>5i5</th>\n",
       "      <th>5i6</th>\n",
       "      <th>6:67</th>\n",
       "      <th>6</th>\n",
       "      <th>6i1</th>\n",
       "      <th>7:66</th>\n",
       "      <th>7</th>\n",
       "      <th>7i1</th>\n",
       "      <th>7i2</th>\n",
       "      <th>7i3</th>\n",
       "      <th>7i4</th>\n",
       "      <th>7i5</th>\n",
       "      <th>7i6</th>\n",
       "      <th>7i7</th>\n",
       "      <th>7i8</th>\n",
       "      <th>7i9</th>\n",
       "      <th>8</th>\n",
       "      <th>8:14:21</th>\n",
       "      <th>8:14</th>\n",
       "      <th>8i1</th>\n",
       "      <th>8i2</th>\n",
       "      <th>9</th>\n",
       "      <th>9:12:23</th>\n",
       "      <th>9:23</th>\n",
       "      <th>9i1</th>\n",
       "      <th>10:25</th>\n",
       "      <th>10</th>\n",
       "      <th>10:25:45</th>\n",
       "      <th>10:45</th>\n",
       "      <th>10i1</th>\n",
       "      <th>11:24</th>\n",
       "      <th>11</th>\n",
       "      <th>12:23</th>\n",
       "      <th>12</th>\n",
       "      <th>12i1</th>\n",
       "      <th>13:22</th>\n",
       "      <th>13</th>\n",
       "      <th>13:22:46</th>\n",
       "      <th>14</th>\n",
       "      <th>14i1</th>\n",
       "      <th>14i2</th>\n",
       "      <th>14i3</th>\n",
       "      <th>14i4</th>\n",
       "      <th>14i5</th>\n",
       "      <th>14i6</th>\n",
       "      <th>14i7</th>\n",
       "      <th>14i8</th>\n",
       "      <th>14i9</th>\n",
       "      <th>14i10</th>\n",
       "      <th>14i11</th>\n",
       "      <th>14i12</th>\n",
       "      <th>14i13</th>\n",
       "      <th>15</th>\n",
       "      <th>15:48</th>\n",
       "      <th>16</th>\n",
       "      <th>16i1</th>\n",
       "      <th>16i2</th>\n",
       "      <th>16i3</th>\n",
       "      <th>16i4</th>\n",
       "      <th>16i5</th>\n",
       "      <th>16i6</th>\n",
       "      <th>16i7</th>\n",
       "      <th>16i8</th>\n",
       "      <th>16i9</th>\n",
       "      <th>16i10</th>\n",
       "      <th>16i11</th>\n",
       "      <th>16i12</th>\n",
       "      <th>16i13</th>\n",
       "      <th>16i14</th>\n",
       "      <th>16i15</th>\n",
       "      <th>16i16</th>\n",
       "      <th>17</th>\n",
       "      <th>17a</th>\n",
       "      <th>18</th>\n",
       "      <th>18:55</th>\n",
       "      <th>19</th>\n",
       "      <th>19:56</th>\n",
       "      <th>19i1</th>\n",
       "      <th>19i2</th>\n",
       "      <th>19i3</th>\n",
       "      <th>19i4</th>\n",
       "      <th>19i5</th>\n",
       "      <th>19i6</th>\n",
       "      <th>19i7</th>\n",
       "      <th>19i8</th>\n",
       "      <th>19i9</th>\n",
       "      <th>19i10</th>\n",
       "      <th>20</th>\n",
       "      <th>20i1</th>\n",
       "      <th>20i2</th>\n",
       "      <th>20i3</th>\n",
       "      <th>20i4</th>\n",
       "      <th>20a</th>\n",
       "      <th>20b</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>22:46</th>\n",
       "      <th>22i1</th>\n",
       "      <th>23</th>\n",
       "      <th>23i1</th>\n",
       "      <th>23i2</th>\n",
       "      <th>23i3</th>\n",
       "      <th>23i4</th>\n",
       "      <th>23i5</th>\n",
       "      <th>23i6</th>\n",
       "      <th>24</th>\n",
       "      <th>24i1</th>\n",
       "      <th>25</th>\n",
       "      <th>25i1</th>\n",
       "      <th>25i2</th>\n",
       "      <th>25i3</th>\n",
       "      <th>25i4</th>\n",
       "      <th>26</th>\n",
       "      <th>26:44</th>\n",
       "      <th>26i1</th>\n",
       "      <th>26i2</th>\n",
       "      <th>26i3</th>\n",
       "      <th>26i4</th>\n",
       "      <th>26i5</th>\n",
       "      <th>26i6</th>\n",
       "      <th>26i7</th>\n",
       "      <th>26i8</th>\n",
       "      <th>26i9</th>\n",
       "      <th>26i10</th>\n",
       "      <th>27:43</th>\n",
       "      <th>27</th>\n",
       "      <th>27i1</th>\n",
       "      <th>27i2</th>\n",
       "      <th>27i3</th>\n",
       "      <th>27i4</th>\n",
       "      <th>28:42</th>\n",
       "      <th>28</th>\n",
       "      <th>28i1</th>\n",
       "      <th>29:41</th>\n",
       "      <th>29</th>\n",
       "      <th>29i1</th>\n",
       "      <th>29i2</th>\n",
       "      <th>30:40</th>\n",
       "      <th>30</th>\n",
       "      <th>30i1</th>\n",
       "      <th>31:39</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>35i1</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>37i1</th>\n",
       "      <th>37i2</th>\n",
       "      <th>37i3</th>\n",
       "      <th>37i4</th>\n",
       "      <th>37i5</th>\n",
       "      <th>37i6</th>\n",
       "      <th>37i7</th>\n",
       "      <th>37i8</th>\n",
       "      <th>37i9</th>\n",
       "      <th>38</th>\n",
       "      <th>38i1</th>\n",
       "      <th>39</th>\n",
       "      <th>39i1</th>\n",
       "      <th>39i2</th>\n",
       "      <th>40</th>\n",
       "      <th>40i1</th>\n",
       "      <th>40i2</th>\n",
       "      <th>41</th>\n",
       "      <th>41i1</th>\n",
       "      <th>41i2</th>\n",
       "      <th>42</th>\n",
       "      <th>42i1</th>\n",
       "      <th>43</th>\n",
       "      <th>44</th>\n",
       "      <th>44i1</th>\n",
       "      <th>44i2</th>\n",
       "      <th>44i3</th>\n",
       "      <th>44i4</th>\n",
       "      <th>44i5</th>\n",
       "      <th>44i6</th>\n",
       "      <th>44i7</th>\n",
       "      <th>44i8</th>\n",
       "      <th>44i9</th>\n",
       "      <th>44i10</th>\n",
       "      <th>44i11</th>\n",
       "      <th>44i12</th>\n",
       "      <th>44i13</th>\n",
       "      <th>44i14</th>\n",
       "      <th>44i15</th>\n",
       "      <th>44i16</th>\n",
       "      <th>44i17</th>\n",
       "      <th>44i18</th>\n",
       "      <th>44i19</th>\n",
       "      <th>44i20</th>\n",
       "      <th>44i21</th>\n",
       "      <th>44i22</th>\n",
       "      <th>44i23</th>\n",
       "      <th>45</th>\n",
       "      <th>V11:V21</th>\n",
       "      <th>V12:V22</th>\n",
       "      <th>V13:V23</th>\n",
       "      <th>V14:V24</th>\n",
       "      <th>V15:V25</th>\n",
       "      <th>V16:V26</th>\n",
       "      <th>V17:V27</th>\n",
       "      <th>V1</th>\n",
       "      <th>V2</th>\n",
       "      <th>V3</th>\n",
       "      <th>V4</th>\n",
       "      <th>V5</th>\n",
       "      <th>V11</th>\n",
       "      <th>V12</th>\n",
       "      <th>V13</th>\n",
       "      <th>V14</th>\n",
       "      <th>V15</th>\n",
       "      <th>V16</th>\n",
       "      <th>V17</th>\n",
       "      <th>V21</th>\n",
       "      <th>V22</th>\n",
       "      <th>V23</th>\n",
       "      <th>V24</th>\n",
       "      <th>V25</th>\n",
       "      <th>V26</th>\n",
       "      <th>V27</th>\n",
       "      <th>46</th>\n",
       "      <th>47</th>\n",
       "      <th>47i1</th>\n",
       "      <th>47i2</th>\n",
       "      <th>47i3</th>\n",
       "      <th>48</th>\n",
       "      <th>49:65</th>\n",
       "      <th>49</th>\n",
       "      <th>49i1</th>\n",
       "      <th>50:64</th>\n",
       "      <th>50</th>\n",
       "      <th>50i1</th>\n",
       "      <th>50i2</th>\n",
       "      <th>50i3</th>\n",
       "      <th>50i4</th>\n",
       "      <th>50i5</th>\n",
       "      <th>50i6</th>\n",
       "      <th>50i7</th>\n",
       "      <th>51:63</th>\n",
       "      <th>51</th>\n",
       "      <th>51i1</th>\n",
       "      <th>51i2</th>\n",
       "      <th>51i3</th>\n",
       "      <th>52:62</th>\n",
       "      <th>52</th>\n",
       "      <th>52i1</th>\n",
       "      <th>53:61</th>\n",
       "      <th>53</th>\n",
       "      <th>53i1</th>\n",
       "      <th>54</th>\n",
       "      <th>54:58</th>\n",
       "      <th>54i1</th>\n",
       "      <th>54i2</th>\n",
       "      <th>54i3</th>\n",
       "      <th>54i4</th>\n",
       "      <th>55</th>\n",
       "      <th>55i1</th>\n",
       "      <th>55i2</th>\n",
       "      <th>55i3</th>\n",
       "      <th>55i4</th>\n",
       "      <th>55i5</th>\n",
       "      <th>56</th>\n",
       "      <th>56i1</th>\n",
       "      <th>56i2</th>\n",
       "      <th>56i3</th>\n",
       "      <th>56i4</th>\n",
       "      <th>57</th>\n",
       "      <th>57i1</th>\n",
       "      <th>58</th>\n",
       "      <th>58i1</th>\n",
       "      <th>58i2</th>\n",
       "      <th>59</th>\n",
       "      <th>59i1</th>\n",
       "      <th>60</th>\n",
       "      <th>60i1</th>\n",
       "      <th>60i2</th>\n",
       "      <th>60i3</th>\n",
       "      <th>60i4</th>\n",
       "      <th>60i5</th>\n",
       "      <th>60i6</th>\n",
       "      <th>60i7</th>\n",
       "      <th>60i8</th>\n",
       "      <th>60i9</th>\n",
       "      <th>60i10</th>\n",
       "      <th>60i11</th>\n",
       "      <th>61</th>\n",
       "      <th>61i1</th>\n",
       "      <th>62</th>\n",
       "      <th>63</th>\n",
       "      <th>64</th>\n",
       "      <th>64i1</th>\n",
       "      <th>64i2</th>\n",
       "      <th>65</th>\n",
       "      <th>65i1</th>\n",
       "      <th>65i2</th>\n",
       "      <th>65i3</th>\n",
       "      <th>65i4</th>\n",
       "      <th>65i5</th>\n",
       "      <th>65i6</th>\n",
       "      <th>65i7</th>\n",
       "      <th>65i8</th>\n",
       "      <th>65i9</th>\n",
       "      <th>66</th>\n",
       "      <th>66i1</th>\n",
       "      <th>67</th>\n",
       "      <th>67i1</th>\n",
       "      <th>68</th>\n",
       "      <th>68i1</th>\n",
       "      <th>68i2</th>\n",
       "      <th>69</th>\n",
       "      <th>69i1</th>\n",
       "      <th>70</th>\n",
       "      <th>70i1</th>\n",
       "      <th>70i2</th>\n",
       "      <th>70i3</th>\n",
       "      <th>70i4</th>\n",
       "      <th>70i5</th>\n",
       "      <th>70i6</th>\n",
       "      <th>70i7</th>\n",
       "      <th>70i8</th>\n",
       "      <th>70i9</th>\n",
       "      <th>71</th>\n",
       "      <th>71i1</th>\n",
       "      <th>71i2</th>\n",
       "      <th>72</th>\n",
       "      <th>73</th>\n",
       "      <th>74</th>\n",
       "      <th>75</th>\n",
       "      <th>76</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Insecta</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Ser</td>\n",
       "      <td>aedAeg1_NW_001809801.1.trna3-SerAGA</td>\n",
       "      <td>aedAeg1</td>\n",
       "      <td>Aedes aegypti (yellow fever mosquito)</td>\n",
       "      <td>.</td>\n",
       "      <td>Arg</td>\n",
       "      <td>AGA</td>\n",
       "      <td>63.6</td>\n",
       "      <td>62.3</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:G:C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:-</td>\n",
       "      <td>G:-</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Insecta</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Ser</td>\n",
       "      <td>aedAeg1_NW_001809801.1.trna4-SerAGA</td>\n",
       "      <td>aedAeg1</td>\n",
       "      <td>Aedes aegypti (yellow fever mosquito)</td>\n",
       "      <td>.</td>\n",
       "      <td>Arg</td>\n",
       "      <td>AGA</td>\n",
       "      <td>63.6</td>\n",
       "      <td>62.3</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:G:C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:-</td>\n",
       "      <td>G:-</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Insecta</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Ser</td>\n",
       "      <td>aedAeg1_NW_001809801.1.trna5-SerAGA</td>\n",
       "      <td>aedAeg1</td>\n",
       "      <td>Aedes aegypti (yellow fever mosquito)</td>\n",
       "      <td>.</td>\n",
       "      <td>Arg</td>\n",
       "      <td>AGA</td>\n",
       "      <td>63.6</td>\n",
       "      <td>62.3</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:G:C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:-</td>\n",
       "      <td>G:-</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Insecta</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Asp</td>\n",
       "      <td>aedAeg1_NW_001809801.1.trna6-AspATC</td>\n",
       "      <td>aedAeg1</td>\n",
       "      <td>Aedes aegypti (yellow fever mosquito)</td>\n",
       "      <td>.</td>\n",
       "      <td>Thr</td>\n",
       "      <td>ATC</td>\n",
       "      <td>70.7</td>\n",
       "      <td>66.7</td>\n",
       "      <td>0.365854</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A:U:A</td>\n",
       "      <td>A:A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:-</td>\n",
       "      <td>G:-</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Insecta</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Ser</td>\n",
       "      <td>aedAeg1_NW_001809801.1.trna15-SerAGA</td>\n",
       "      <td>aedAeg1</td>\n",
       "      <td>Aedes aegypti (yellow fever mosquito)</td>\n",
       "      <td>.</td>\n",
       "      <td>Arg</td>\n",
       "      <td>AGA</td>\n",
       "      <td>60.2</td>\n",
       "      <td>61.5</td>\n",
       "      <td>0.341463</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:G:C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:-</td>\n",
       "      <td>G:-</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     clade     domain isotype                               seqname  species                           species_long taxid isotype_best anticodon  score  isoscore        GC  intron  insertions  deletions  D-loop  AC-loop  TPC-loop  V-arm quality restrict 1:72  1 1i1 2:71  2 2i1 3:70  3 3i1 3i2 3i3 3i4 3i5 3i6 4:69  4 4i1 4i2 4i3 4i4 4i5 4i6 4i7 4i8 4i9 4i10 4i11 5:68  5 5i1 5i2 5i3 5i4 5i5 5i6 6:67  6 6i1 7:66  7 7i1 7i2 7i3 7i4 7i5 7i6 7i7 7i8 7i9  8 8:14:21 8:14 8i1 8i2  9 9:12:23 9:23 9i1 10:25 10 10:25:45 10:45 10i1 11:24 11 12:23 12 12i1 13:22 13 13:22:46 14 14i1 14i2 14i3 14i4 14i5 14i6 14i7 14i8 14i9 14i10 14i11 14i12 14i13 15 15:48 16 16i1 16i2 16i3 16i4 16i5 16i6 16i7 16i8 16i9 16i10 16i11 16i12 16i13 16i14 16i15 16i16 17 17a 18 18:55 19 19:56 19i1 19i2 19i3 19i4 19i5 19i6 19i7 19i8 19i9 19i10 20 20i1 20i2 20i3 20i4 20a 20b 21 22 22:46 22i1 23 23i1 23i2 23i3 23i4 23i5 23i6 24 24i1 25 25i1 25i2 25i3 25i4 26 26:44 26i1 26i2 26i3 26i4 26i5 26i6 26i7 26i8 26i9 26i10 27:43 27 27i1 27i2 27i3 27i4 28:42 28 28i1 29:41 29 29i1 29i2 30:40 30 30i1 31:39 31 32 33 34 35 35i1 36 37 37i1 37i2 37i3 37i4 37i5 37i6 37i7 37i8 37i9 38 38i1 39 39i1 39i2 40 40i1 40i2 41 41i1 41i2 42 42i1 43 44 44i1 44i2 44i3 44i4 44i5 44i6 44i7 44i8 44i9 44i10 44i11 44i12 44i13 44i14 44i15 44i16 44i17 44i18 44i19 44i20 44i21 44i22 44i23 45 V11:V21 V12:V22 V13:V23 V14:V24 V15:V25 V16:V26 V17:V27 V1 V2 V3 V4 V5 V11 V12 V13 V14 V15 V16 V17 V21 V22 V23 V24 V25 V26 V27 46 47 47i1 47i2 47i3 48 49:65 49 49i1 50:64 50 50i1 50i2 50i3 50i4 50i5 50i6 50i7 51:63 51 51i1 51i2 51i3 52:62 52 52i1 53:61 53 53i1 54 54:58 54i1 54i2 54i3 54i4 55 55i1 55i2 55i3 55i4 55i5 56 56i1 56i2 56i3 56i4 57 57i1 58 58i1 58i2 59 59i1 60 60i1 60i2 60i3 60i4 60i5 60i6 60i7 60i8 60i9 60i10 60i11 61 61i1 62 63 64 64i1 64i2 65 65i1 65i2 65i3 65i4 65i5 65i6 65i7 65i8 65i9 66 66i1 67 67i1 68 68i1 68i2 69 69i1 70 70i1 70i2 70i3 70i4 70i5 70i6 70i7 70i8 70i9 71 71i1 71i2 72 73 74 75 76\n",
       "0  Insecta  eukaryota     Ser   aedAeg1_NW_001809801.1.trna3-SerAGA  aedAeg1  Aedes aegypti (yellow fever mosquito)     .          Arg       AGA   63.6      62.3  0.390244       0           2          1      10        7         8      0   False    False  C:G  C   .  C:A  C   .  U:A  U   .   .   .   .   .   .  C:G  C   .   .   .   .   .   .   .   .   .    .    .  U:A  U   .   .   .   .   .   .  C:G  C   .  A:U  A   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:G:C  G:C   .   G:C  G    G:C:-   G:-    .   C:G  C   G:C  G    .   U:A  U    U:A:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   U   -  A  A   A:G    .  C    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   C:G  C    .   C:G  C    .    .   C:G  C    A   A:U  A  C  U  A  G    .  A  G    .    .    .    .    .    .    .    .    .  A    .  U    .    C  G    .    .  G    .    .  G    .  G  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  -     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  G  U    .    .    .  C   G:C  G    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   A:U  A    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  U    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  U  C  A    .    .  C    .    .    .    .    .    .    .    .    .  U    .  G    .  A    .    .  G    .  A    .    .    .    .    .    .    .    .    .  A    .    .  G  A  -  -  -\n",
       "1  Insecta  eukaryota     Ser   aedAeg1_NW_001809801.1.trna4-SerAGA  aedAeg1  Aedes aegypti (yellow fever mosquito)     .          Arg       AGA   63.6      62.3  0.390244       0           2          1      10        7         8      0   False     True  C:G  C   .  C:A  C   .  U:A  U   .   .   .   .   .   .  C:G  C   .   .   .   .   .   .   .   .   .    .    .  U:A  U   .   .   .   .   .   .  C:G  C   .  A:U  A   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:G:C  G:C   .   G:C  G    G:C:-   G:-    .   C:G  C   G:C  G    .   U:A  U    U:A:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   U   -  A  A   A:G    .  C    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   C:G  C    .   C:G  C    .    .   C:G  C    A   A:U  A  C  U  A  G    .  A  G    .    .    .    .    .    .    .    .    .  A    .  U    .    C  G    .    .  G    .    .  G    .  G  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  -     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  G  U    .    .    .  C   G:C  G    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   A:U  A    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  U    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  U  C  A    .    .  C    .    .    .    .    .    .    .    .    .  U    .  G    .  A    .    .  G    .  A    .    .    .    .    .    .    .    .    .  A    .    .  G  A  -  -  -\n",
       "2  Insecta  eukaryota     Ser   aedAeg1_NW_001809801.1.trna5-SerAGA  aedAeg1  Aedes aegypti (yellow fever mosquito)     .          Arg       AGA   63.6      62.3  0.390244       0           2          1      10        7         8      0   False     True  C:G  C   .  C:A  C   .  U:A  U   .   .   .   .   .   .  C:G  C   .   .   .   .   .   .   .   .   .    .    .  U:A  U   .   .   .   .   .   .  C:G  C   .  A:U  A   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:G:C  G:C   .   G:C  G    G:C:-   G:-    .   C:G  C   G:C  G    .   U:A  U    U:A:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   U   -  A  A   A:G    .  C    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   C:G  C    .   C:G  C    .    .   C:G  C    A   A:U  A  C  U  A  G    .  A  G    .    .    .    .    .    .    .    .    .  A    .  U    .    C  G    .    .  G    .    .  G    .  G  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  -     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  G  U    .    .    .  C   G:C  G    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   A:U  A    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  U    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  U  C  A    .    .  C    .    .    .    .    .    .    .    .    .  U    .  G    .  A    .    .  G    .  A    .    .    .    .    .    .    .    .    .  A    .    .  G  A  -  -  -\n",
       "3  Insecta  eukaryota     Asp   aedAeg1_NW_001809801.1.trna6-AspATC  aedAeg1  Aedes aegypti (yellow fever mosquito)     .          Thr       ATC   70.7      66.7  0.365854       0           0          1       8        7         8      0   False    False  U:A  U   .  G:C  G   .  C:G  C   .   .   .   .   .   .  U:A  U   .   .   .   .   .   .   .   .   .    .    .  C:G  C   .   .   .   .   .   .  G:C  G   .  G:C  G   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  A   A:U:A  A:A   .   G:C  G    G:C:-   G:-    .   C:G  C   U:A  U    .   U:A  U    U:A:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  U   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   -   -  A  A   A:G    .  A    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   U:A  U    .   C:A  C    .    .   G:C  G    .   U:A  U  C  U  A  U    .  C  A    .    .    .    .    .    .    .    .    .  U    .  A    .    .  C    .    .  A    .    .  A    .  G  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  -     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  G  U    .    .    .  C   C:G  C    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   G:C  G    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  A    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  C  C  A    .    .  G    .    .    .    .    .    .    .    .    .  C    .  C    .  G    .    .  A    .  G    .    .    .    .    .    .    .    .    .  C    .    .  A  C  -  -  -\n",
       "4  Insecta  eukaryota     Ser  aedAeg1_NW_001809801.1.trna15-SerAGA  aedAeg1  Aedes aegypti (yellow fever mosquito)     .          Arg       AGA   60.2      61.5  0.341463       0           2          1       9        7         8      0   False     True  C:G  C   .  C:A  C   .  U:A  U   .   .   .   .   .   .  C:G  C   .   .   .   .   .   .   .   .   .    .    .  U:A  U   .   .   .   .   .   .  C:G  C   .  A:U  A   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:G:C  G:C   .   G:C  G    G:C:-   G:-    .   C:G  C   G:C  G    .   U:A  U    U:A:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   -   -  A  A   A:G    .  C    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   C:G  C    C   U:A  U    .    .   A:C  A    .   A:U  A  C  U  A  G    .  A  G    .    .    .    .    .    .    .    .    .  A    .  U    .    .  C    .    .  A    .    G  G    .  G  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  -     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  G  U    .    .    .  C   G:C  G    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   A:U  A    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  U    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  U  C  A    .    .  C    .    .    .    .    .    .    .    .    .  U    .  G    .  A    .    .  G    .  A    .    .    .    .    .    .    .    .    .  A    .    .  G  A  -  -  -"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "identities.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Resolve consensus positions\n",
    "\n",
    "There are some discrepencies between the covariance model, the Sprinzl model, and reality. For instance, is position 17 really a gap position for most tRNAs? If so, why is it considered a position? We know that the D loop interacts with the TPC loop, so the general position 17 area should be a key identity element or structural feature contributing to tRNA activity. What should we call insertions at 17 if it's not a real position? These conflicts need to be resolved into a consensus version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def bounds_to_cols(cols, start, end):\n",
    "  selected_cols = []\n",
    "  for col in cols:\n",
    "    matches = re.findall('\\d+', col)\n",
    "    if len(matches) < 1: continue\n",
    "    index = int(matches[0])\n",
    "    if (index >= start and index <= end or col[0:3] == '{}i'.format(start - 1)) and col[0] != 'V':\n",
    "      selected_cols.append(col)\n",
    "  return selected_cols\n",
    "\n",
    "def position_histogram(identities, cols):\n",
    "  df = pd.melt(identities[cols + ['seqname']], id_vars=['seqname'])\n",
    "  df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "  counts = pd.Series(Counter([x for x in df.variable]), index=cols)\n",
    "  counts = counts[counts > 10]\n",
    "  counts = pd.DataFrame({\"Position\": counts.index, 'tRNAs': counts})\n",
    "  return counts.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### D-loop positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dloop_cols = bounds_to_cols(identities.columns, 14, 21)\n",
    "position_histogram(identities, dloop_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the model curated using --hand, the D loop now contains 17a, 20a, and 20b. The alignment seems to have been fixed, as we don't have a bump at 20i5 (or even 20i5-20i7) anymore.\n",
    "\n",
    "#### 17A\n",
    "\n",
    "What isotypes have position 17?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(identities[['isotype', '17']], id_vars=['isotype'])\n",
    "df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "df = pd.Series(Counter([x for x in df.isotype]))\n",
    "df.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(identities[['isotype', '17a']], id_vars=['isotype'])\n",
    "df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "df = pd.Series(Counter([x for x in df.isotype]))\n",
    "df.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on Sprinzl alignments, there are 19 tRNAs out of 2249 with a base at position 17A. It also doesn't look like we can hardcode position 17 for just one or two isotypes - it's universal enough. Based on this data, 17A is pretty rare, and even position 17 is only there 30% of the time (based on our data). Position 17A could actually be considered as an insertion.\n",
    "\n",
    "#### 20a/20b\n",
    "\n",
    "** This section is somewhat deprecated as the alignment has been fixed. ** The following analysis has been modified to double check whether the new 20a/20b alignment is sound.\n",
    "\n",
    "Also based on Sprinzl alignments, no tRNAs have only 20b and not 20a. So whenever there is 20b, 20a is sure to be there. Let's take a look at the distribution out of 2249 tRNA genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sprinzl_20ab = pd.read_table('sprinzl-20ab.txt', header=None, names=['20a', '20b', 'Isotype'])\n",
    "sprinzl_20ab = sprinzl_20ab.groupby([\"Isotype\"]).aggregate(lambda x: len(x[x != '-']))\n",
    "sprinzl_20ab.plot(kind=\"bar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This complicates things. Ideally we'd like to see all of the tRNAs with 20b to belong to one or two isotypes. Then we could manually check the most common insertion positions for those isotypes and adjust it. Instead, we'll just have to check the insertion positions for all isotypes and attempt to come up with some universal rules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sprinzl_20ab = pd.read_table('sprinzl-20ab.txt', header=None, names=['20a', '20b', 'Isotype'])\n",
    "sprinzl_20ab = sprinzl_20ab.groupby([\"20a\", \"20b\"]).aggregate(len)\n",
    "sprinzl_20ab.plot(kind=\"bar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This plot tells us a few things.\n",
    "\n",
    "First, 20a is typically a T, but there are a few ACG 20as. To be expected, as 20ab are typically modified to dihydrouridine (D).\n",
    "\n",
    "20b is typically also a T, but is most commonly paired with C20a. There are a few of the others, too. [Nissan and Perona (2000)](https://www.ncbi.nlm.nih.gov/pubmed/11105758) corroborate this by showing a successful replacement of  yeast function serine U20a/G20b with C20a/A20b _in vitro_. This complicates things. \n",
    "\n",
    "I hesitate to use a rule like \"First base is 20a, second base is 20b, third and up is 20i\". There is high potential for misannotation of an insertion as 20ab or vice versa. We can agree on the following:\n",
    "- single base insertion: 20a\n",
    "- two base insertion: 20a, 20b\n",
    "\n",
    "For three and up, we could potentially leave it unannotated. Let's take a look at the relative insertion lengths, by isotype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "insert_df = identities[['isotype'] + ['20a', '20b', '20i1', '20i2', '20i3', '20i4']]\n",
    "insertionab = insert_df.ix[:, 1:].apply(lambda x: sum((x != '.') & (x != '-')), axis=1)\n",
    "insert_df = insert_df.assign(insertionab = insertionab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(insert_df.groupby('insertionab').aggregate({\"isotype\": len}))\n",
    "plot = sns.violinplot(data=insert_df, x='isotype', y='insertionab', scale='width', cut=0)\n",
    "plot.set(ylim=(-1, 5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Half of the tRNAs have no 20a/20b, and the other half does. For 10 isotypes, all tRNAs are 2 insertions and below. For the other 10, it varies. Exactly how many tRNAs are we working with here? Is there a correlation betweeen number of tRNAs and number of tRNAs with insertions > 2?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df = insert_df.ix[insert_df.insertionab > 2, ['isotype', 'insertionab']].groupby('isotype').aggregate(len)\n",
    "print(\"No. tRNAs with > 2 insertions: {}\".format(sum(df.insertionab)))\n",
    "df = df.merge(identities[['isotype', 'seqname']].groupby('isotype').aggregate(len), how='right', right_index=True, left_index=True)\n",
    "df = df.fillna(0)\n",
    "df.columns = ['insertionab_tRNAs', 'tRNAs']\n",
    "\n",
    "# scatter plot\n",
    "plot = df.plot(kind='scatter', x='insertionab_tRNAs', y='tRNAs')\n",
    "for isotype, num_tRNAs in df.iterrows(): plot.annotate(isotype, num_tRNAs, xytext=(5, 0), textcoords='offset points', family='sans-serif')\n",
    "\n",
    "# add trendline\n",
    "beta = numpy.polyfit(x=df.insertionab_tRNAs, y=df.tRNAs, deg=1)\n",
    "trendline = pd.DataFrame({'x': list(df.insertionab_tRNAs) + [-20, 120], 'trendline': numpy.poly1d(beta)(list(df.insertionab_tRNAs) + [-20, 120])})\n",
    "trendline.plot(ax=plot, x='x', y='trendline', kind='line', xlim=[-3, 10], ylim=[0, 2500])\n",
    "\n",
    "plot.set_xlabel('No. tRNAs with >2 insertions at position 20')\n",
    "plot.set_ylabel('No. tRNAs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There's only 38 tRNAs with a 3 base or larger insertion. Here they are below. They're mostly leucine tRNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = insert_df[insert_df.insertionab > 2]\n",
    "df[\"seq\"] = df[['20a', '20b', '20i1', '20i2', '20i3', '20i4']].apply(lambda x: ''.join(x[(x != '.') & (x != '-')]), axis=1)\n",
    "df.sort_values('insertionab')[['isotype', 'seq']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variable stem-loop plus spacer nucleotides\n",
    "\n",
    "** This section is somewhat deprecated as the alignment has been fixed. ** The following analysis has been modified to double check whether the position 47 and V-arm alignment is sound."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "varm_cols = list(filter(lambda x: ('V' in x) & (':' not in x), identities.columns))\n",
    "varm_cols += bounds_to_cols(identities.columns, 44, 48)\n",
    "position_histogram(identities, varm_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data correlates well with the Sprinzl data. Most variable stems are missing at least V15:V25, V16:V26, and V17:V27, and most variable loops are 3 nucleotides long.\n",
    "\n",
    "Here is the same plot, but split into Ser, Leu, and Tyr, versus everything else."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(identities[varm_cols + ['seqname', 'isotype']], id_vars=['seqname', 'isotype'])\n",
    "df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "counts = pd.Series(Counter([x for x in df[(df.isotype == \"Tyr\") | (df.isotype == \"Leu\") | (df.isotype == \"Ser\")].variable]), index=varm_cols)\n",
    "counts = pd.DataFrame({\"Position\": counts.index, 'tRNAs': counts})\n",
    "counts2 = pd.Series(Counter([x for x in df[(df.isotype != \"Tyr\") | (df.isotype != \"Leu\") | (df.isotype != \"Ser\")].variable]), index=varm_cols)\n",
    "counts = pd.merge(counts, pd.DataFrame({\"Position\": counts2.index, 'tRNAs': counts2}), on=['Position'])\n",
    "counts.columns = ['Position', 'Class II', 'Class I']\n",
    "counts.index = counts.Position\n",
    "counts.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to apply a similar analysis for 47 as we did for 20ab. What does the distribution of 47 + insert length look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "insert_df = identities[['isotype', '47', '47i1']]\n",
    "insert_df.head()\n",
    "insertion47 = 2 - sum([(insert_df.ix[:, 1] == \".\"), (insert_df.ix[:, 2] == \".\"), (insert_df.ix[:, 1] == \"-\"), (insert_df.ix[:, 2] == \"-\")])\n",
    "insert_df = insert_df.assign(insertion47 = insertion47)\n",
    "print(insert_df.groupby('insertion47').aggregate({\"isotype\": len}))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of the 94197 tRNAs with 1 nucleotide at position 47 + 47ix, which position does it appear on? Is there a typical position 47 nucleotide?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(insert_df.ix[insert_df.insertion47 == 1, 0:3], id_vars='isotype')\n",
    "df = df[(df.value != \"-\") & (df.value != '.')]\n",
    "df[['variable', 'value']].groupby(['variable', 'value']).aggregate(len).unstack().plot(kind='bar')\n",
    "df.ix[df.variable != \"47\", :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 0 cases where a single base is not aligned to position 47 properly."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
