{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import subprocess\n",
    "import glob\n",
    "import re\n",
    "import os, sys\n",
    "from collections import defaultdict, Counter, OrderedDict\n",
    "from Bio import SeqIO, SeqRecord, Seq\n",
    "from tRNA_position import *\n",
    "pd.set_option('display.max_colwidth',10000)\n",
    "pd.set_option('display.width', 10000)\n",
    "pd.set_option('display.max_columns', 10000)\n",
    "pd.set_option('display.max_rows',1000)\n",
    "isotypes = ['Ala', 'Arg', 'Asn', 'Asp', 'Cys', 'Gln', 'Glu', 'Gly', 'His', 'Ile', 'iMet', 'Leu', 'Lys', 'Met', 'Phe', 'Pro', 'Ser', 'Thr', 'Trp', 'Tyr', 'Val']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "A global view of identity elements versus biological features would be a powerful tool for predicting tRNA function using primary sequence. To do this, I'll need to align eukaryotic tRNAs to our established models and annotate positions based on universal numbering. Then, I'll extract sequence information such as position, clade, or variable arm length."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process tRNAs\n",
    "## Species information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>longname</th>\n",
       "      <th>domain</th>\n",
       "      <th>clade</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ashbGoss_ATCC10895</td>\n",
       "      <td>Eremothecium gossypii ATCC 10895</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>33169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aspeFumi_AF293</td>\n",
       "      <td>Aspergillus fumigatus Af293</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>746128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aspeNidu_FGSC_A4</td>\n",
       "      <td>Aspergillus nidulans FGSC A4</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>162425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aspeOryz_RIB40</td>\n",
       "      <td>Aspergillus oryzae RIB40</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>5062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>botrCine_B05_10</td>\n",
       "      <td>Botrytis cinerea B05.10</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>40559</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              species                          longname     domain  clade   taxid\n",
       "0  ashbGoss_ATCC10895  Eremothecium gossypii ATCC 10895  eukaryota  Fungi   33169\n",
       "1      aspeFumi_AF293       Aspergillus fumigatus Af293  eukaryota  Fungi  746128\n",
       "2    aspeNidu_FGSC_A4      Aspergillus nidulans FGSC A4  eukaryota  Fungi  162425\n",
       "3      aspeOryz_RIB40          Aspergillus oryzae RIB40  eukaryota  Fungi    5062\n",
       "4     botrCine_B05_10           Botrytis cinerea B05.10  eukaryota  Fungi   40559"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_table = pd.read_table('genomes-051418.tsv', header=None, names=['species', 'longname', 'domain', 'clade', 'taxid'], dtype={'taxid': str})\n",
    "species_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align tRNAs to Sprinzl numbering model\n",
    "\n",
    "We have a \"quality set\" of tRNAs, which include tRNAs from previous tRNAscan-SE runs and newer runs. We also want to keep track of high-scoring, isotype mismatched tRNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing ailMel1...finished\n",
      "processing anoCar2...finished\n",
      "processing anoGam2...finished\n",
      "processing apiMel1...finished\n",
      "processing araTha1...finished\n",
      "processing ashbGoss_ATCC10895...finished\n",
      "processing aspeFumi_AF293...finished\n",
      "processing aspeNidu_FGSC_A4...finished\n",
      "processing aspeOryz_RIB40...finished\n",
      "processing balAcu1...finished\n",
      "processing bomTer1...finished\n",
      "processing bosTau8...finished\n",
      "processing botrCine_B05_10...finished\n",
      "processing braDis2...finished\n",
      "processing braOle1...finished\n",
      "processing bruMal1...finished\n",
      "processing caeJap1...finished\n",
      "processing caePb2...finished\n",
      "processing caeRem3...finished\n",
      "processing calJac3...finished\n",
      "processing calMil1...finished\n",
      "processing canFam3...finished\n",
      "processing candAlbi_WO_1...finished\n",
      "processing candDubl_CD36...finished\n",
      "processing candGlab_CBS_138...finished\n",
      "processing candOrth_CO_90_125...finished\n",
      "processing carPap1...finished\n",
      "processing cavPor3...finished\n",
      "processing cb3...finished\n",
      "processing ce11...finished\n",
      "processing cerSim1...finished\n",
      "processing chaeTher_VAR_THERMOPHILUM_DSM1...finished\n",
      "processing chrPic1...finished\n",
      "processing criGri1...finished\n",
      "processing crypGatt_WM276...finished\n",
      "processing crypNeof_VAR_GRUBII_H99...finished\n",
      "processing crypNeof_VAR_NEOFORMANS_B_3501...finished\n",
      "processing crypNeof_VAR_NEOFORMANS_JEC21...finished\n",
      "processing cucSat1...finished\n",
      "processing cybeJadi_NBRC_0988_NBRC0988...finished\n",
      "processing dasNov3...finished\n",
      "processing debaHans_CBS767...finished\n",
      "processing dipOrd1...finished\n",
      "processing dm6...finished\n",
      "processing dp4...finished\n",
      "processing droAna3...finished\n",
      "processing droEre2...finished\n",
      "processing droGri2...finished\n",
      "processing droMoj3...finished\n",
      "processing droPer1...finished\n",
      "processing droSec1...finished\n",
      "processing droSim1...finished\n",
      "processing droVir3...finished\n",
      "processing droWil1...finished\n",
      "processing droYak2...finished\n",
      "processing enceCuni_GB_M1...finished\n",
      "processing enceHell_ATCC50504...finished\n",
      "processing enceInte_ATCC50506...finished\n",
      "processing enceRoma_SJ_2008...finished\n",
      "processing equCab2...finished\n",
      "processing eremCymb_DBVPG#7215...finished\n",
      "processing eriEur2...finished\n",
      "processing felCat5...finished\n",
      "processing flamVelu_KACC42780...finished\n",
      "processing fr3...finished\n",
      "processing fraVes1...finished\n",
      "processing fusaGram_CS3005...finished\n",
      "processing fusaGram_PH_1_NRRL_31084...finished\n",
      "processing fusaPseu_CS3096...finished\n",
      "processing fusaPseu_CS3270...finished\n",
      "processing fusaVert_7600...finished\n",
      "processing gadMor1...finished\n",
      "processing galGal5...finished\n",
      "processing gasAcu1...finished\n",
      "processing geoFor1...finished\n",
      "processing glyMax2...finished\n",
      "processing gorGor4...finished\n",
      "processing gosRai2...finished\n",
      "processing hetBac1...finished\n",
      "processing hetGla2...finished\n",
      "processing hetGly1...finished\n",
      "processing hg19...finished\n",
      "processing kazaAfri_CBS_2517...finished\n",
      "processing kazaNaga_CBS_8797...finished\n",
      "processing kluyLact_NRRL_Y_1140...finished\n",
      "processing komaPast_CBS_7435...finished\n",
      "processing komaPast_GS115...finished\n",
      "processing lachKluy_NRRL_Y_12651...finished\n",
      "processing lachTher_CBS_6340...finished\n",
      "processing latCha1...finished\n",
      "processing loaLoa1...finished\n",
      "processing loxAfr3...finished\n",
      "processing macEug2...finished\n",
      "processing magnOryz_70_15...finished\n",
      "processing malDom1...finished\n",
      "processing manEsc6...finished\n",
      "processing mayDes1...finished\n",
      "processing medTru4...finished\n",
      "processing melGal5...finished\n",
      "processing melUnd1...finished\n",
      "processing micMur2...finished\n",
      "processing millFari_CBS_7064...finished\n",
      "processing mm10...finished\n",
      "processing monDom5...finished\n",
      "processing musFur1...finished\n",
      "processing myceTher_ATCC42464...finished\n",
      "processing myoLuc2...finished\n",
      "processing nasVit1...finished\n",
      "processing naumCast_CBS_4309...finished\n",
      "processing naumDair_CBS_421...finished\n",
      "processing neurCras_OR74A...finished\n",
      "processing nomLeu3...finished\n",
      "processing ochPri3...finished\n",
      "processing ogatPara_DL_1...finished\n",
      "processing oreNil2...finished\n",
      "processing ornAna1...finished\n",
      "processing oryCun2...finished\n",
      "processing oryLat2...finished\n",
      "processing orySat7...finished\n",
      "processing otoGar3...finished\n",
      "processing oviAri3...finished\n",
      "processing panRed1...finished\n",
      "processing panTro4...finished\n",
      "processing papAnu2...finished\n",
      "processing peniChry_P2NIAD18...finished\n",
      "processing petMar2...finished\n",
      "processing phyPat3...finished\n",
      "processing ponAbe2...finished\n",
      "processing popTri3...finished\n",
      "processing priPac1...finished\n",
      "processing proCap1...finished\n",
      "processing pteVam1...finished\n",
      "processing rheMac8...finished\n",
      "processing rn6...finished\n",
      "processing sacCer3...finished\n",
      "processing saccSp_ASHBYA_ACERI...finished\n",
      "processing saccSp_BOULARDII_17...finished\n",
      "processing saccSp_BOULARDII_ATCCMYA_796...finished\n",
      "processing saiBol1...finished\n",
      "processing sarHar1...finished\n",
      "processing scheStip_CBS_6054...finished\n",
      "processing schiPomb_972H...finished\n",
      "processing selMoe1...finished\n",
      "processing solPen1...finished\n",
      "processing sorAra2...finished\n",
      "processing sorBic3...finished\n",
      "processing speTri2...finished\n",
      "processing spiPol1...finished\n",
      "processing sporReil_SRZ2...finished\n",
      "processing strRat1...finished\n",
      "processing susScr11...finished\n",
      "processing taeGut2...finished\n",
      "processing tarSyr2...finished\n",
      "processing tetNig2...finished\n",
      "processing tetrBlat_CBS_6284...finished\n",
      "processing tetrPhaf_CBS_4417...finished\n",
      "processing thieTerr_NRRL_8126...finished\n",
      "processing toruDelb_CBS_1146...finished\n",
      "processing triCas1...finished\n",
      "processing triMan1...finished\n",
      "processing triSpi1...finished\n",
      "processing turTru2..."
     ]
    }
   ],
   "source": [
    "species = sorted(glob.glob(\"out/*-detailed.out\"))\n",
    "species = [sp[4:-19] for sp in species]\n",
    "seqs = []\n",
    "for sp in species:\n",
    "  if sp not in species_table.values: continue\n",
    "  clade = species_table.loc[species_table.species == sp, 'clade'].values[0]\n",
    "  sys.stdout.write('processing ' + sp + '...')\n",
    "  sys.stdout.flush()\n",
    "  \n",
    "  tRNA_file = 'tRNAs/' + sp + '-tRNAs.fa'\n",
    "  tscanout_file = 'out/{}-tRNAs-detailed.out'.format(sp)\n",
    "  # Import annotation from tRNAscan .out files.\n",
    "  # Filter tRNAs here to speed up data import. We want to keep a small number of lower quality tRNAs in case it's needed for lifting to other isotypes.\n",
    "  approved_tRNAs = []\n",
    "  intron_lengths = [] # we are purging introns to improve alignment, so store intron lengths here\n",
    "  for metadata in pd.read_table(tscanout_file, sep=\"\\t\", skiprows=3, na_filter = False, header=None).iterrows():\n",
    "    if ((clade == 'Fungi' and metadata[1].iloc[-1] == \"Isotype mismatch;Isotype mismatch;First-pass quality filtered\") or\n",
    "         metadata[1].iloc[-1] in [\"\",\n",
    "                                  \"high confidence set\",\n",
    "                                  \"secondary filtered\",\n",
    "                                  \"tertiary filtered\", \n",
    "                                  \"Quality set\",\n",
    "                                  \"unexpected anticodon\",\n",
    "                                  \"Unexpected anticodon;First-pass quality filtered\",\n",
    "                                  \"Isotype mismatch;First-pass quality filtered\",\n",
    "                                  \"First-pass quality filtered\",\n",
    "                                  \"Second-pass quality filtered\"] or\n",
    "         re.match('^IPD:-\\d+\\.\\d+(,(isotype mismatch|secondary filtered|tertiary filtered|unexpected anticodon))?$', metadata[1].iloc[-1]) or\n",
    "         (clade == 'Fungi' and 'cryp' in sp and metadata[1].iloc[4] == 'Asp' and float(metadata[1].iloc[8]) > 35)):\n",
    "      if metadata[1][4] == 'Met' and metadata[1][12] == 'iMet': isotype = 'iMet'\n",
    "      else: isotype = metadata[1][4]\n",
    "      approved_tRNAs.append('{}.trna{}-{}{}'.format(str(metadata[1][0]).strip(), metadata[1][1], isotype, metadata[1][5]))\n",
    "      intron_length = abs(int(metadata[1][6]) - int(metadata[1][7]))\n",
    "      if intron_length > 0: intron_length = intron_length + 1\n",
    "      intron_lengths.append(intron_length)\n",
    "  # Remove introns using sstofa\n",
    "  subprocess.call('sstofa3 ss/{}-tRNAs.ss \"\" 1 0 > {}'.format(sp, tRNA_file), shell=True)\n",
    "\n",
    "  # Parse isotype-specific scores file\n",
    "  iso_scores = pd.read_table('iso/' + sp + '-tRNAs.iso', header=0)\n",
    "  iso_scores['Undet'] = 0\n",
    "  iso_scores['Sup'] = 0\n",
    "  iso_scores['best'] = iso_scores.iloc[:,2:].idxmax(axis=1)\n",
    "  iso_scores['score'] = iso_scores.max(axis=1, numeric_only=True)\n",
    "  iso_scores['ac_score'] = iso_scores.lookup(iso_scores.index, iso_scores.iloc[:, 1])\n",
    "  iso_scores.index = iso_scores.tRNAscanID.values\n",
    "  iso_scores = iso_scores[['best', 'score', 'ac_score']]\n",
    "  \n",
    "  # Parse seqs, filtering out the ones we don't need\n",
    "  for seq in SeqIO.parse('tRNAs/{}-tRNAs.fa'.format(sp), 'fasta'):\n",
    "    if seq.id not in approved_tRNAs:\n",
    "      seq.id = seq.id.replace('Met', 'iMet')\n",
    "      if seq.id not in approved_tRNAs: continue\n",
    "    if \"pseudogene\" in seq.description and 'cryp' not in sp: continue\n",
    "    intron_length = intron_lengths[approved_tRNAs.index(seq.id)]\n",
    "    score = float(re.findall('Sc: [\\d\\.]+', seq.description)[0].split()[-1])\n",
    "    # relax tRNA score requirements for fungi\n",
    "    if (species_table[species_table.species == sp].clade.values[0] == \"Fungi\"):\n",
    "      if score < 30 and seq.description != 'chrI.trna9-MetCAT (3515186-3515314)  Met (CAT) 70 bp  Sc: 27.6': continue\n",
    "    elif species_table[species_table.species == sp].clade.values[0] == \"Insect\":\n",
    "      if score < 50: continue\n",
    "    else:\n",
    "      if score < 55: continue\n",
    "    trnascanid = re.findall('.+\\.trna\\d+', seq.id)[0]\n",
    "    best_isotype = iso_scores.loc[trnascanid].best\n",
    "    if best_isotype == 'SeC' or 'mito' in best_isotype or 'mt' in trnascanid: continue\n",
    "    isoscore, ac_score = iso_scores.loc[trnascanid].score, iso_scores.loc[trnascanid].ac_score\n",
    "    seq.id = '{}|{} Iso: {} ({}) Iso_ac: {} Intron: {}'.format(sp, seq.description, isoscore, best_isotype, ac_score, intron_length)\n",
    "    seq.description = ''\n",
    "    seqs.append(seq)\n",
    "\n",
    "  print('finished')\n",
    "  sys.stdout.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_handle = open('euk-tRNAs.fa', 'w')\n",
    "SeqIO.write(seqs, fasta_handle, 'fasta')\n",
    "fasta_handle.close()\n",
    "num_model = '/projects/lowelab/users/blin/tRNAscan/models/domain-specific/euk-num-092016.cm'\n",
    "subprocess.call('cmalign -g --notrunc -o euk-tRNAs.sto {} euk-tRNAs.fa'.format(num_model), shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create table of tRNA bases by position\n",
    "\n",
    "This is a giant data frame with one row per tRNA, and with columns for each position, plus tRNA metadata like species and loop lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def position_base(positions, seq):\n",
    "  for position_index, position in enumerate(positions):\n",
    "    if position.paired:\n",
    "      index1, index2 = position.position.split(':')\n",
    "      index1, index2 = int(index1), int(index2)\n",
    "      base_pair = \"{}:{}\".format(seq[index1 - 1].upper(), seq[index2 - 1].upper())\n",
    "      yield position.sprinzl, base_pair\n",
    "    else:\n",
    "      index = int(position.position)\n",
    "      base = seq[index - 1].upper()\n",
    "      yield position.sprinzl, base\n",
    "      \n",
    "identities = pd.DataFrame()\n",
    "\n",
    "# get positions\n",
    "alignment_fhandle = open('euk-tRNAs.sto')\n",
    "positions = [] # list containing each position in the tRNA\n",
    "\n",
    "# first, get secondary structure\n",
    "for line in alignment_fhandle:\n",
    "  if line[0:12] == '#=GC SS_cons':\n",
    "    ss = line.strip().split()[-1]\n",
    "alignment_fhandle.close()\n",
    "\n",
    "# parse secondary structure into regions and positions\n",
    "positions = annotate_positions(ss)\n",
    "\n",
    "# get nucleotide at each position for each tRNA by parsing Stockholm file\n",
    "alignment_fhandle = open('euk-tRNAs.sto')\n",
    "trnas = []\n",
    "skipped = []\n",
    "for line in alignment_fhandle:\n",
    "  if line[0] in [\"#\", '\\n', '/']: continue\n",
    "  species, desc = line.strip().split('|', 1)\n",
    "  seqname = desc.split()[0]\n",
    "  seq = desc.split()[-1]\n",
    "  if any(species_table.species == species):\n",
    "    row = species_table[species_table.species == species]\n",
    "  else: \n",
    "    skipped.append(species)\n",
    "    continue\n",
    "  isotype = re.findall('\\.trna\\d+-([A-Za-z]+)', seqname)[0][:-3]\n",
    "  seqname = '{}_{}'.format(species, seqname)\n",
    "  trna = {'domain': row.domain.values[0], 'clade': row.clade.values[0], 'species': species, 'species_long': row.longname.values[0], 'taxid': row.taxid.values[0], 'seqname': seqname, 'isotype': isotype}\n",
    "  trna = {**trna, **{sprinzl: base for sprinzl, base in position_base(positions, seq)}}\n",
    "  trnas.append(trna)\n",
    "\n",
    "if len(skipped) > 0: print('skipped the following: {}'.format(set(skipped)))\n",
    "identities = identities.append(trnas, ignore_index=True)\n",
    "identities.fillna('.', inplace=True)\n",
    "alignment_fhandle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create single base columns from paired positions\n",
    "\n",
    "This is mainly for plotting and such."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = list(filter(lambda x: ':' in x, identities.columns))\n",
    "for col in cols:\n",
    "  pos1, pos2 = col.split(':')\n",
    "  base1 = [bases.split(':')[0] for bases in identities[col]]\n",
    "  base2 = [bases.split(':')[1] for bases in identities[col]]\n",
    "  identities[pos1] = base1\n",
    "  identities[pos2] = base2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate tertiary interactions\n",
    "\n",
    "We also want to annotate the 2-base version of the base triples. This makes checking for conserved interactions (e.g. an identity element) a bit easier. It's important to remember to consider the 3rd base pair too, though."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "identities['8:14:21'] = identities.loc[:, [\"8\", \"14\", \"21\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['8:14'] = identities.loc[:, [\"8\", \"14\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['9:12:23'] = identities.loc[:, [\"9\", \"12\", \"23\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['9:23'] = identities.loc[:, [\"9\", \"23\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['10:25:45'] = identities.loc[:, [\"10\", \"25\", \"45\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['10:45'] = identities.loc[:, [\"10\", \"45\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['13:22:46'] = identities.loc[:, [\"13\", \"22\", \"46\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['22:46'] = identities.loc[:, [\"22\", \"46\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['15:48'] = identities.loc[:, [\"15\", \"48\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['18:55'] = identities.loc[:, [\"18\", \"55\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['19:56'] = identities.loc[:, [\"19\", \"56\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['26:44'] = identities.loc[:, [\"26\", \"44\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['54:58'] = identities.loc[:, [\"54\", \"58\"]].apply(lambda row: ':'.join(row), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional sequence information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic information - isotype, anticodon, score\n",
    "seqinfo = []\n",
    "for line in open('euk-tRNAs.sto'):\n",
    "  if line[0:4] != \"#=GS\": continue\n",
    "  _, desc, _, _, isotype, anticodon, _, _, _, score, _, isoscore, isotype_best, _, isoscore_ac, _, intron_length = line.strip().split()\n",
    "  seqname = desc.replace('|', '_', 1)\n",
    "  seqinfo.append([seqname, isotype_best[1:-1], anticodon[1:-1], float(score), float(isoscore), float(isoscore_ac), int(intron_length)])\n",
    "seqinfo = pd.DataFrame(seqinfo, columns=['seqname', 'isotype_best', 'anticodon', 'score', 'isoscore', 'isoscore_ac', 'intron'])\n",
    "identities = identities.merge(seqinfo, on='seqname')\n",
    "\n",
    "# GC content\n",
    "paired_cols = identities.columns[list(map(lambda x: (':' in x), identities.columns))]\n",
    "identities['GC'] = identities[paired_cols].apply(lambda x: sum((x == \"G:C\") | (x == \"C:G\"))/len(paired_cols), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adjust isotype designations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# iMet may or may not be properly annotated, depending on tRNAscan-SE run\n",
    "unlifted_imets = identities[(identities.isotype == \"Met\") & (identities.isotype_best == \"iMet\")].index\n",
    "identities.loc[unlifted_imets, 'isotype'] = 'iMet'\n",
    "\n",
    "identities['adjusted'] = False\n",
    "# Manage tRNAs of missing isotypes\n",
    "for sp in identities.species.unique():\n",
    "  # Lift the highest-scoring iMet if species does not contain iMet\n",
    "  if 'iMet' not in identities[identities.species == sp].isotype.values:\n",
    "    # reparse iso file and re-assign the Met with the highest iMet score\n",
    "    imet_scores = {}\n",
    "    seqname_isotypes = {}\n",
    "    for line in open('iso/{}-tRNAs.iso'.format(sp)):\n",
    "      tabs = line.strip().split('\\t')\n",
    "      if tabs[0] == 'tRNAscanID': continue\n",
    "      seqname = sp + '_' + tabs[0]\n",
    "      imet_scores[seqname] = float(tabs[-1])\n",
    "      seqname_isotypes[seqname] = tabs[1]\n",
    "    \n",
    "    best_seqname = max(imet_scores, key = imet_scores.get)\n",
    "    best_score = np.max(imet_scores.values())\n",
    "    print('{}: adjusting {} to iMet'.format(sp, identities.loc[identities.seqname.apply(lambda x: x[:-7]) == best_seqname, 'seqname'].values.tolist()))\n",
    "    identities.loc[identities.seqname.apply(lambda x: x[:-7]) == best_seqname, 'isotype'] = 'iMet'\n",
    "    identities.loc[identities.seqname.apply(lambda x: x[:-7]) == best_seqname, 'adjusted'] = True\n",
    "    for seqname in imet_scores:\n",
    "      if abs(imet_scores[seqname] - imet_scores[best_seqname]) < 2 and seqname_isotypes[seqname] == seqname_isotypes[best_seqname]:\n",
    "        print('{}: adjusting {} to iMet'.format(sp, identities.loc[identities.seqname.apply(lambda x: x[:-7]) == seqname, 'seqname'].values.tolist()))\n",
    "        identities.loc[identities.seqname.apply(lambda x: x[:-7]) == seqname, 'isotype'] = 'iMet'\n",
    "        identities.loc[identities.seqname.apply(lambda x: x[:-7]) == seqname, 'adjusted'] = True\n",
    "\n",
    "  # Revert tRNAs in species that are missing tRNAs of an isotype\n",
    "  for isotype in isotypes:\n",
    "    if isotype not in identities.loc[identities.species == sp].isotype.unique():\n",
    "      if identities.loc[(identities.species == sp) & (identities.seqname.apply(lambda x: isotype in x))].seqname.size > 0:\n",
    "        print('{}: adjusting {} to {}'.format(sp, identities.loc[(identities.species == sp) & (identities.seqname.apply(lambda x: isotype in x)), 'seqname'].values.tolist(), isotype))\n",
    "        identities.loc[(identities.species == sp) & (identities.seqname.apply(lambda x: isotype in x)), 'isotype'] = isotype\n",
    "        identities.loc[(identities.species == sp) & (identities.seqname.apply(lambda x: isotype in x)), 'adjusted'] = True\n",
    "      else:\n",
    "        print('{}: missing {} tRNAs'.format(sp, isotype))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insertions/deletions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Insertions (minus misaligned introns at 37/38)\n",
    "intron_cols = list(filter(lambda x: x[0:3] == '37i', identities.columns))\n",
    "insertion_cols = list(filter(lambda x: bool(re.search('^\\d+i', x)) & (x not in intron_cols), identities.columns))\n",
    "identities['insertions'] = identities[insertion_cols].apply(lambda x: sum(x != '.'), axis=1)\n",
    "\n",
    "# Deletions at positions that are not the variable arm, and not counting 17/17a/20a/20b\n",
    "base_cols = list(filter(lambda x: bool(re.match('^\\d+$', x)) & (x not in ['74', '75', '76', '17', '17a', '20a', '20b']), identities.columns))\n",
    "identities['deletions'] = identities[base_cols].apply(lambda x: ''.join(x).count('-'), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loop sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def bounds_to_cols(cols, start, end):\n",
    "  selected_cols = []\n",
    "  for col in cols:\n",
    "    matches = re.findall('\\d+', col)\n",
    "    if len(matches) < 1: continue\n",
    "    index = int(matches[0])\n",
    "    if (index >= start and index <= end or col[0:3] == '{}i'.format(start - 1)) and col[0] != 'V':\n",
    "      selected_cols.append(col)\n",
    "  return selected_cols\n",
    "\n",
    "dloop_cols = list(filter(lambda col: ':' not in col, bounds_to_cols(identities.columns, 14, 21)))\n",
    "identities['D-loop'] = identities[dloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "# Leu, Ser have a 3 bp D stem\n",
    "dloop_II_cols = list(filter(lambda col: ':' not in col, bounds_to_cols(identities.columns, 13, 22)))\n",
    "identities.loc[(identities.isotype == 'Leu') | (identities.isotype == 'Ser'), 'D-loop'] = identities[dloop_II_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "acloop_cols = list(filter(lambda x: not re.match('37i.+', x), bounds_to_cols(identities.columns, 32, 38)))\n",
    "identities['AC-loop'] = identities[acloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "tpcloop_cols = bounds_to_cols(identities.columns, 54, 60)\n",
    "identities['TPC-loop'] = identities[tpcloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "varm_cols = list(filter(lambda x: ('V' in x) & (':' not in x) | (x in bounds_to_cols(identities.columns, 45, 48)), identities.columns))\n",
    "identities['V-arm'] = identities[varm_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Restrict tRNAs by species\n",
    "\n",
    "We may also want to limit the contribution of any single species, similar to how we built the isotype-specific models. tRNAs that are restricted are excluded from analysis. We'll start from the confidence set if applicable and add in unique tRNAs from there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Import list of quality tRNAs\n",
    "confidence_trnas = [line.strip() for line in open('confidence-set.txt')]\n",
    "identities['confidence'] = identities.seqname.isin(confidence_trnas)\n",
    "\n",
    "identities['restrict'] = True\n",
    "identities['duplicate'] = True\n",
    "\n",
    "identities.loc[identities.adjusted, 'restrict'] = False\n",
    "identities.loc[identities.score > 45, 'restrict'] = False\n",
    "\n",
    "for species in identities.species.unique():\n",
    "  for isotype in isotypes:\n",
    "    isotype_indices = identities.loc[(identities.species == species) & (identities.isotype == isotype) & (-identities.restrict)].index\n",
    "    confidence_indices = identities.loc[(identities.species == species) & (identities.isotype == isotype) & (-identities.restrict) & (identities.confidence)].index\n",
    "    unique_scores_indices = isotype_indices[-identities.iloc[isotype_indices].score.duplicated()]\n",
    "    unique_confidence_indices = confidence_indices[-identities.iloc[confidence_indices].score.duplicated()]\n",
    "    \n",
    "    # no tRNAs means this is a fungi missing an isotype, lower score threshold to 30 and skip rest\n",
    "    if len(unique_scores_indices) == 0: \n",
    "      isotype_indices = identities.loc[(identities.species == species) & (identities.isotype == isotype)].index\n",
    "      if (len(isotype_indices) == 0): continue\n",
    "      unique_scores_indices = isotype_indices[-identities.iloc[isotype_indices].score.duplicated()]\n",
    "      identities.loc[unique_scores_indices, 'duplicate'] = False\n",
    "      sorted_unique_indices = unique_scores_indices[np.argsort(identities.loc[unique_scores_indices, 'score'])]\n",
    "      identities.loc[sorted_unique_indices[-1], 'adjusted'] = True\n",
    "      identities.loc[sorted_unique_indices[-1], 'restrict'] = False\n",
    "      identities.loc[sorted_unique_indices[:-1], 'restrict'] = True\n",
    "      continue\n",
    "      \n",
    "    identities.loc[unique_scores_indices, 'duplicate'] = False\n",
    "    \n",
    "    if len(unique_confidence_indices) >= 50: # too many unique tRNAs; likely repetitive\n",
    "      unique_confidence_indices = unique_confidence_indices[np.argsort(identities.loc[unique_confidence_indices, 'score'])][::-1][:50]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_confidence_indices))\n",
    "    elif len(unique_confidence_indices) > 0:\n",
    "      if len(unique_scores_indices) > 50 - len(unique_confidence_indices):\n",
    "        unique_scores_indices = unique_scores_indices[np.argsort(identities.loc[unique_scores_indices, 'score'])][::-1][:50 - len(unique_confidence_indices)]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_confidence_indices) - set(unique_scores_indices))\n",
    "    else:\n",
    "      if len(unique_scores_indices) > 50:\n",
    "        unique_scores_indices = unique_scores_indices[np.argsort(identities.loc[unique_scores_indices, 'score'])][::-1][:50]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_scores_indices))\n",
    "    \n",
    "    identities.loc[restricted_indices, 'restrict'] = True\n",
    "\n",
    "identities.loc[identities.duplicate, 'restrict'] = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The quality set does have some issues, though: it filters out rare codons, when rare is relative to model species. It also performs a hard check for anticodon-isotype model match. Fungi have much more diverged tRNAs and, in many cases, a different isotype scores higher even when it's clear that it's the bona fide tRNA (mostly based on number of tRNAs of that isotype, or codon usage). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export to R\n",
    "\n",
    "R has superior visualization capabilities.\n",
    "\n",
    "### Order columns\n",
    "\n",
    "To make it look pretty."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def position_str_to_int(position):\n",
    "  if position == \"20a\": return 20.1\n",
    "  if position == \"20b\": return 20.2\n",
    "  digits = re.findall('\\d+', position)\n",
    "  if len(digits) == 0: return -1\n",
    "  insert = 0\n",
    "  if 'i' in position and len(digits) == 2: insert = float(digits[1]) / 1000\n",
    "  if position[0] == 'V':\n",
    "    if ':' in position: return int(digits[0]) + 45 - 10 + insert # V11~V17\n",
    "    else: return int(digits[0]) + 45 + 7 + insert # V1~V5\n",
    "  if int(digits[0]) >= 46: return int(digits[0]) + 50 + insert # just add an arbitrarily large number to skip v-arm\n",
    "  return int(digits[0]) + insert\n",
    "\n",
    "identities = identities[sorted(list(identities.columns), key=position_str_to_int)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "identities.to_csv(path_or_buf='identities.tsv', sep='\\t', index_label=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export fasta of tRNAs used for analysis\n",
    "fasta_handle = open('euk-tRNAs-filtered.fa', 'w')\n",
    "filtered_seqs = []\n",
    "seqnames = identities[~identities.restrict].seqname.tolist()\n",
    "for seq in SeqIO.parse('euk-tRNAs.fa', 'fasta'):\n",
    "  seqname = seq.name.replace('|', '_')\n",
    "  if seqname in seqnames: filtered_seqs.append(seq)\n",
    "SeqIO.write(filtered_seqs, fasta_handle, 'fasta')\n",
    "fasta_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "identities.head()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
