{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from ggplot import *\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "mpl.rcParams['figure.figsize'] = (15, 6)\n",
    "mpl.style.use('ggplot')\n",
    "import pandas as pd\n",
    "import numpy\n",
    "import subprocess\n",
    "import glob\n",
    "import re\n",
    "import os, sys\n",
    "from collections import defaultdict, Counter, OrderedDict\n",
    "from Bio import SeqIO, SeqRecord, Seq\n",
    "from tRNA_position import *\n",
    "pd.set_option('display.max_colwidth',10000)\n",
    "pd.set_option('display.width', 10000)\n",
    "pd.set_option('display.max_columns', 10000)\n",
    "pd.set_option('display.max_rows',1000)\n",
    "isotypes = ['Ala', 'Arg', 'Asn', 'Asp', 'Cys', 'Gln', 'Glu', 'Gly', 'His', 'Ile', 'iMet', 'Leu', 'Lys', 'Met', 'Phe', 'Pro', 'Ser', 'Thr', 'Trp', 'Tyr', 'Val']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "A global view of identity elements versus biological features would be a powerful tool for predicting tRNA function using primary sequence. To do this, I'll need to align eukaryotic tRNAs to our established models and annotate positions based on universal numbering. Then, I'll extract sequence information such as position, clade, or variable arm length."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process tRNAs\n",
    "## Species information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>longname</th>\n",
       "      <th>domain</th>\n",
       "      <th>clade</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ashbGoss_ATCC10895</td>\n",
       "      <td>Eremothecium gossypii ATCC 10895</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>33169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aspeFumi_AF293</td>\n",
       "      <td>Aspergillus fumigatus Af293</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>746128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aspeNidu_FGSC_A4</td>\n",
       "      <td>Aspergillus nidulans FGSC A4</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>162425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>aspeOryz_RIB40</td>\n",
       "      <td>Aspergillus oryzae RIB40</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>5062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>botrCine_B05_10</td>\n",
       "      <td>Botrytis cinerea B05.10</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>40559</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              species                          longname     domain  clade   taxid\n",
       "0  ashbGoss_ATCC10895  Eremothecium gossypii ATCC 10895  eukaryota  Fungi   33169\n",
       "1      aspeFumi_AF293       Aspergillus fumigatus Af293  eukaryota  Fungi  746128\n",
       "2    aspeNidu_FGSC_A4      Aspergillus nidulans FGSC A4  eukaryota  Fungi  162425\n",
       "3      aspeOryz_RIB40          Aspergillus oryzae RIB40  eukaryota  Fungi    5062\n",
       "4     botrCine_B05_10           Botrytis cinerea B05.10  eukaryota  Fungi   40559"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_table = pd.read_table('genomes-061517.tsv', header=None, names=['species', 'longname', 'domain', 'clade', 'taxid'], dtype={'taxid': str})\n",
    "species_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align tRNAs to Sprinzl numbering model\n",
    "\n",
    "We have a \"quality set\" of tRNAs, which include tRNAs from previous tRNAscan-SE runs and newer runs. We also want to keep track of high-scoring, isotype mismatched tRNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing ailMel1...finished\n",
      "processing anoCar2...finished\n",
      "processing anoGam2...finished\n",
      "processing apiMel1...finished\n",
      "processing araTha1...finished\n",
      "processing ashbGoss_ATCC10895...finished\n",
      "processing aspeFumi_AF293...finished\n",
      "processing aspeNidu_FGSC_A4...finished\n",
      "processing aspeOryz_RIB40...finished\n",
      "processing balAcu1...finished\n",
      "processing bomTer1...finished\n",
      "processing bosTau8...finished\n",
      "processing botrCine_B05_10...finished\n",
      "processing braDis3...finished\n",
      "processing braOle1...finished\n",
      "processing bruMal1...finished\n",
      "processing caeJap1...finished\n",
      "processing caePb2...finished\n",
      "processing caeRem3...finished\n",
      "processing calJac3...finished\n",
      "processing calMil1...finished\n",
      "processing canFam3...finished\n",
      "processing candAlbi_WO_1...finished\n",
      "processing candDubl_CD36...finished\n",
      "processing candGlab_CBS_138...finished\n",
      "processing candOrth_CO_90_125...finished\n",
      "processing carPap1...finished\n",
      "processing cavPor3...finished\n",
      "processing cb3...finished\n",
      "processing ce11...finished\n",
      "processing cerSim1...finished\n",
      "processing chaeTher_VAR_THERMOPHILUM_DSM1...finished\n",
      "processing chrPic1...finished\n",
      "processing criGri1...finished\n",
      "processing crypGatt_WM276...finished\n",
      "processing crypNeof_VAR_GRUBII_H99...finished\n",
      "processing crypNeof_VAR_NEOFORMANS_B_3501...finished\n",
      "processing crypNeof_VAR_NEOFORMANS_JEC21...finished\n",
      "processing cucSat1...finished\n",
      "processing cybeJadi_NBRC_0988_NBRC0988...finished\n",
      "processing danRer10...finished\n",
      "processing dasNov3...finished\n",
      "processing debaHans_CBS767...finished\n",
      "processing dipOrd1...finished\n",
      "processing dm6...finished\n",
      "processing dp4...finished\n",
      "processing droAna3...finished\n",
      "processing droEre2...finished\n",
      "processing droGri2...finished\n",
      "processing droMoj3...finished\n",
      "processing droPer1...finished\n",
      "processing droSec1...finished\n",
      "processing droSim1...finished\n",
      "processing droVir3...finished\n",
      "processing droWil1...finished\n",
      "processing droYak2...finished\n",
      "processing enceCuni_GB_M1...finished\n",
      "processing enceHell_ATCC50504...finished\n",
      "processing enceInte_ATCC50506...finished\n",
      "processing enceRoma_SJ_2008...finished\n",
      "processing equCab2...finished\n",
      "processing eremCymb_DBVPG#7215...finished\n",
      "processing eriEur2...finished\n",
      "processing felCat5...finished\n",
      "processing flamVelu_KACC42780...finished\n",
      "processing fr3...finished\n",
      "processing fraVes1...finished\n",
      "processing fusaGram_CS3005...finished\n",
      "processing fusaGram_PH_1_NRRL_31084...finished\n",
      "processing fusaPseu_CS3096...finished\n",
      "processing fusaPseu_CS3270...finished\n",
      "processing fusaVert_7600...finished\n",
      "processing gadMor1...finished\n",
      "processing galGal4...finished\n",
      "processing gasAcu1...finished\n",
      "processing geoFor1...finished\n",
      "processing glyMax2...finished\n",
      "processing gorGor3...finished\n",
      "processing gosRai2...finished\n",
      "processing hetBac1...finished\n",
      "processing hetGla2...finished\n",
      "processing hetGly1...finished\n",
      "processing hg19...finished\n",
      "processing kazaAfri_CBS_2517...finished\n",
      "processing kazaNaga_CBS_8797...finished\n",
      "processing kluyLact_NRRL_Y_1140...finished\n",
      "processing komaPast_CBS_7435...finished\n",
      "processing komaPast_GS115...finished\n",
      "processing lachKluy_NRRL_Y_12651...finished\n",
      "processing lachTher_CBS_6340...finished\n",
      "processing latCha1...finished\n",
      "processing loaLoa1...finished\n",
      "processing loxAfr3...finished\n",
      "processing macEug2...finished\n",
      "processing magnOryz_70_15...finished\n",
      "processing malDom1...finished\n",
      "processing manEsc6...finished\n",
      "processing mayDes1...finished\n",
      "processing medTru4...finished\n",
      "processing melGal1...finished\n",
      "processing melUnd1...finished\n",
      "processing micMur1...finished\n",
      "processing millFari_CBS_7064...finished\n",
      "processing mm10...finished\n",
      "processing monDom5...finished\n",
      "processing musFur1...finished\n",
      "processing myceTher_ATCC42464...finished\n",
      "processing myoLuc2...finished\n",
      "processing nasVit1...finished\n",
      "processing naumCast_CBS_4309...finished\n",
      "processing naumDair_CBS_421...finished\n",
      "processing neurCras_OR74A...finished\n",
      "processing nomLeu3...finished\n",
      "processing ochPri3...finished\n",
      "processing ogatPara_DL_1...finished\n",
      "processing oreNil2...finished\n",
      "processing ornAna1...finished\n",
      "processing oryCun2...finished\n",
      "processing oryLat2...finished\n",
      "processing orySat7...finished\n",
      "processing otoGar3...finished\n",
      "processing oviAri3...finished\n",
      "processing panRed1...finished\n",
      "processing panTro4...finished\n",
      "processing papAnu2...finished\n",
      "processing peniChry_P2NIAD18...finished\n",
      "processing ponAbe2...finished\n",
      "processing popTri3...finished\n",
      "processing priPac1...finished\n",
      "processing proCap1...finished\n",
      "processing pteVam1...finished\n",
      "processing rheMac8...finished\n",
      "processing rn6...finished\n",
      "processing sacCer3...finished\n",
      "processing saccCere_AWRI796...finished\n",
      "processing saccCere_CEN_PK113_7D...finished\n",
      "processing saccCere_FOSTERSB...finished\n",
      "processing saccCere_P283...finished\n",
      "processing saccCere_P301...finished\n",
      "processing saccCere_R008...finished\n",
      "processing saccCere_R103...finished\n",
      "processing saccCere_SIGMA1278B...finished\n",
      "processing saccCere_UFMG_A_905...finished\n",
      "processing saccCere_VIN13...finished\n",
      "processing saccCere_VL3...finished\n",
      "processing saccCere_W303...finished\n",
      "processing saccCere_YJSH1...finished\n",
      "processing saccSp_ASHBYA_ACERI...finished\n",
      "processing saccSp_BOULARDII_17...finished\n",
      "processing saccSp_BOULARDII_ATCCMYA_796...finished\n",
      "processing saiBol1...finished\n",
      "processing sarHar1...finished\n",
      "processing scheStip_CBS_6054...finished\n",
      "processing schiPomb_972H...finished\n",
      "processing solPen1...finished\n",
      "processing sorAra2...finished\n",
      "processing sorBic3...finished\n",
      "processing speTri2...finished\n",
      "processing spiPol1...finished\n",
      "processing sporReil_SRZ2...finished\n",
      "processing strRat1...finished\n",
      "processing susScr3...finished\n",
      "processing taeGut2...finished\n",
      "processing tarSyr2...finished\n",
      "processing tetNig2...finished\n",
      "processing tetrBlat_CBS_6284...finished\n",
      "processing tetrPhaf_CBS_4417...finished\n",
      "processing thieTerr_NRRL_8126...finished\n",
      "processing timCri1...finished\n",
      "processing toruDelb_CBS_1146...finished\n",
      "processing triCas1...finished\n",
      "processing triMan1...finished\n",
      "processing triSpi1...finished\n",
      "processing turTru2...finished\n",
      "processing ustiMayd_521...finished\n",
      "processing valsMali_03_8...finished\n",
      "processing vertDahl_JR2...finished\n",
      "processing vertDahl_VDLS_17...finished\n",
      "processing vicPac2...finished\n",
      "processing vitVin1...finished\n",
      "processing xenTro3...finished\n",
      "processing yarrLipo_CLIB122...finished\n",
      "processing yarrLipo_PO1F...finished\n",
      "processing yarrLipo_WSH_Z06...finished\n",
      "processing zeaMay6...finished\n",
      "processing zygoRoux_CBS_732...finished\n"
     ]
    }
   ],
   "source": [
    "species = sorted(glob.glob(\"out/*-detailed.out\"))\n",
    "species = [sp[4:-19] for sp in species]\n",
    "seqs = []\n",
    "for sp in species:  \n",
    "  sys.stdout.write('processing ' + sp + '...')\n",
    "  sys.stdout.flush()\n",
    "  \n",
    "  tRNA_file = 'tRNAs/' + sp + '-tRNAs.fa'\n",
    "  tscanout_file = 'out/{}-tRNAs-detailed.out'.format(sp)\n",
    "  # Import annotation from tRNAscan .out files. We want to keep the ones that are \"Quality set\" or \"Unexpected anticodon;First-pass quality filtered\".\n",
    "  approved_tRNAs = []\n",
    "  for metadata in pd.read_table(tscanout_file, sep=\"\\t\", skiprows=3, header=None).iterrows():\n",
    "    if (metadata[1].iloc[-1] in [\"Quality set\", \"Unexpected anticodon;First-pass quality filtered\", \"quality set\", \"unexpected anticodon,first-pass quality filtered\", \"Isotype mismatch;First-pass quality filtered\"]\n",
    "        or re.match('^IPD:-\\d+\\.\\d+,isotype mismatch,first-pass quality filtered$', metadata[1].iloc[-1])):\n",
    "      approved_tRNAs.append('{}.trna{}-{}{}'.format(metadata[1][0], metadata[1][1], metadata[1][4], metadata[1][5]))\n",
    " \n",
    "  # Remove introns using sstofa\n",
    "  subprocess.call('sstofa3 ss/{}-tRNAs.ss \"\" 1 0 > {}'.format(sp, tRNA_file), shell=True)\n",
    "\n",
    "  # Parse isotype-specific scores file\n",
    "  iso_scores = pd.read_table('iso/' + sp + '-tRNAs.iso', header=0)\n",
    "  iso_scores['Undet'] = 0\n",
    "  iso_scores['Sup'] = 0\n",
    "  iso_scores['best'] = iso_scores.ix[:,2:].idxmax(axis=1)\n",
    "  iso_scores['score'] = iso_scores.max(axis=1, numeric_only=True)\n",
    "  iso_scores['ac_score'] = iso_scores.lookup(iso_scores.index, iso_scores.ix[:, 1])\n",
    "  iso_scores.index = iso_scores.tRNAscanID.values\n",
    "  iso_scores = iso_scores[['best', 'score', 'ac_score']]\n",
    "  \n",
    "  # Parse seqs, filtering out the ones we don't need, and add to proper isotype\n",
    "  for seq in SeqIO.parse('tRNAs/{}-tRNAs.fa'.format(sp), 'fasta'):\n",
    "    if seq.id not in approved_tRNAs: continue\n",
    "    if \"pseudogene\" in seq.description: continue\n",
    "    score = float(re.findall('Sc: [\\d\\.]+', seq.description)[0].split()[-1])\n",
    "    if score < 55: continue\n",
    "    trnascanid = re.findall('.+\\.trna\\d+', seq.id)[0]\n",
    "    isotype = iso_scores.ix[trnascanid].best\n",
    "    if isotype == 'SeC' or 'mito' in isotype or 'mt' in trnascanid: continue\n",
    "    isoscore, ac_score = iso_scores.ix[trnascanid].score, iso_scores.ix[trnascanid].ac_score\n",
    "    seq.id = '{}|{} Iso: {} ({}) Iso_ac: {}'.format(sp, seq.description, isoscore, isotype, ac_score)\n",
    "    seq.description = ''\n",
    "    seqs.append(seq)\n",
    "    \n",
    "  print('finished')\n",
    "  sys.stdout.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_handle = open('euk-tRNAs.fa', 'w')\n",
    "SeqIO.write(seqs, fasta_handle, 'fasta')\n",
    "fasta_handle.close()\n",
    "num_model = '/projects/lowelab/users/blin/tRNAscan/models/domain-specific/euk-num-092016.cm'\n",
    "subprocess.call('cmalign -g --notrunc -o euk-tRNAs.sto {} euk-tRNAs.fa'.format(num_model), shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create table of tRNA bases by position\n",
    "\n",
    "This is a giant data frame with one row per tRNA, and with columns for each position, plus tRNA metadata like species and loop lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "skipped the following: {'saccCere_FOSTERSB', 'saccCere_P283', 'saccCere_P301', 'saccCere_SIGMA1278B', 'saccCere_R103', 'saccCere_UFMG_A_905', 'saccCere_CEN_PK113_7D', 'saccCere_VIN13', 'saccCere_R008', 'saccCere_VL3', 'saccCere_YJSH1', 'saccCere_AWRI796', 'saccCere_W303'}\n"
     ]
    }
   ],
   "source": [
    "def position_base(positions, seq):\n",
    "  for position_index, position in enumerate(positions):\n",
    "    if position.paired:\n",
    "      index1, index2 = position.position.split(':')\n",
    "      index1, index2 = int(index1), int(index2)\n",
    "      base_pair = \"{}:{}\".format(seq[index1 - 1].upper(), seq[index2 - 1].upper())\n",
    "      yield position.sprinzl, base_pair\n",
    "    else:\n",
    "      index = int(position.position)\n",
    "      base = seq[index - 1].upper()\n",
    "      yield position.sprinzl, base\n",
    "      \n",
    "identities = pd.DataFrame()\n",
    "\n",
    "# get positions\n",
    "alignment_fhandle = open('euk-tRNAs.sto')\n",
    "positions = [] # list containing each position in the tRNA\n",
    "\n",
    "# first, get secondary structure\n",
    "for line in alignment_fhandle:\n",
    "  if line[0:12] == '#=GC SS_cons':\n",
    "    ss = line.strip().split()[-1]\n",
    "alignment_fhandle.close()\n",
    "\n",
    "# parse secondary structure into regions and positions\n",
    "positions = annotate_positions(ss)\n",
    "\n",
    "# get nucleotide at each position for each tRNA by parsing Stockholm file\n",
    "alignment_fhandle = open('euk-tRNAs.sto')\n",
    "trnas = []\n",
    "skipped = []\n",
    "for line in alignment_fhandle:\n",
    "  if line[0] in [\"#\", '\\n', '/']: continue\n",
    "  species, desc = line.strip().split('|', 1)\n",
    "  seqname = desc.split()[0]\n",
    "  seq = desc.split()[-1]\n",
    "  if any(species_table.species == species):\n",
    "    row = species_table[species_table.species == species]\n",
    "  else: \n",
    "    skipped.append(species)\n",
    "    continue\n",
    "  isotype = re.findall('\\.trna\\d+-([A-Za-z]+)', seqname)[0][:-3]\n",
    "  seqname = '{}_{}'.format(species, seqname)\n",
    "  trna = {'domain': row.domain.values[0], 'clade': row.clade.values[0], 'species': species, 'species_long': row.longname.values[0], 'taxid': row.taxid.values[0], 'seqname': seqname, 'isotype': isotype}\n",
    "  trna = {**trna, **{sprinzl: base for sprinzl, base in position_base(positions, seq)}}\n",
    "  trnas.append(trna)\n",
    "\n",
    "if len(skipped) > 0: print('skipped the following: {}'.format(set(skipped)))\n",
    "identities = identities.append(trnas, ignore_index=True)\n",
    "identities.fillna('.', inplace=True)\n",
    "alignment_fhandle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create single base columns from paired positions\n",
    "\n",
    "This is mainly for plotting and such."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cols = list(filter(lambda x: ':' in x, identities.columns))\n",
    "for col in cols:\n",
    "  pos1, pos2 = col.split(':')\n",
    "  base1 = [bases.split(':')[0] for bases in identities[col]]\n",
    "  base2 = [bases.split(':')[1] for bases in identities[col]]\n",
    "  identities[pos1] = base1\n",
    "  identities[pos2] = base2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Shorten D stem for class II tRNAs\n",
    "\n",
    "Leu and Tyr have a 3 base pair D stem instead of 4. Since we've already copied over the columns into single bases, this is okay. I don't want to change the entries for this position to \"-:-\" because it would imply a deletion, and I don't want to change it to \".:.\" because it would break a lot of code. I'll leave it alone for now, until it's necessary to change it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate tertiary interactions\n",
    "\n",
    "We also want to annotate the 2-base version of the base triples. This makes checking for conserved interactions (e.g. an identity element) a bit easier. It's important to remember to consider the 3rd base pair too, though."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "identities['8:14:21'] = identities.ix[:, [\"8\", \"14\", \"21\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['8:14'] = identities.ix[:, [\"8\", \"14\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['9:12:23'] = identities.ix[:, [\"9\", \"12\", \"23\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['9:23'] = identities.ix[:, [\"9\", \"23\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['10:25:45'] = identities.ix[:, [\"10\", \"25\", \"45\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['10:45'] = identities.ix[:, [\"10\", \"45\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['13:22:46'] = identities.ix[:, [\"13\", \"22\", \"46\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['22:46'] = identities.ix[:, [\"22\", \"46\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['15:48'] = identities.ix[:, [\"15\", \"48\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['18:55'] = identities.ix[:, [\"18\", \"55\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['19:56'] = identities.ix[:, [\"19\", \"56\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['26:44'] = identities.ix[:, [\"26\", \"44\"]].apply(lambda row: ':'.join(row), axis=1)\n",
    "identities['54:58'] = identities.ix[:, [\"54\", \"58\"]].apply(lambda row: ':'.join(row), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional sequence information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Isotype, anticodon, score, isoscores\n",
    "seqinfo = []\n",
    "for line in open('euk-tRNAs.sto'):\n",
    "  if line[0:4] != \"#=GS\": continue\n",
    "  _, desc, _, _, isotype, anticodon, _, _, _, score, _, isoscore, isotype_best, _, isoscore_ac = line.strip().split()\n",
    "  seqname = desc.replace('|', '_', 1)\n",
    "  seqinfo.append([seqname, isotype_best[1:-1], anticodon[1:-1], float(score), float(isoscore), float(isoscore_ac)])\n",
    "seqinfo = pd.DataFrame(seqinfo, columns=['seqname', 'isotype_best', 'anticodon', 'score', 'isoscore', 'isoscore_ac'])\n",
    "identities = identities.merge(seqinfo, on='seqname')\n",
    "\n",
    "# iMet may or may not be properly annotated, depending on tRNAscan-SE run\n",
    "unlifted_imets = identities.ix[(identities.isotype == \"Met\") & (identities.isotype_best == \"iMet\"), ].index\n",
    "identities.ix[unlifted_imets, 'isotype'] = 'iMet'\n",
    "\n",
    "# GC content\n",
    "paired_cols = identities.columns[list(map(lambda x: (':' in x), identities.columns))]\n",
    "identities['GC'] = identities[paired_cols].apply(lambda x: sum((x == \"G:C\") | (x == \"C:G\"))/len(paired_cols), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insertions/deletions/introns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Intron length\n",
    "intron_cols = list(filter(lambda x: x[0:3] == '37i', identities.columns))\n",
    "identities['intron'] = identities[intron_cols].apply(lambda x: sum(x != \".\"), axis=1)\n",
    "\n",
    "# Insertions (minus introns at 37/38)\n",
    "insertion_cols = list(filter(lambda x: bool(re.search('^\\d+i', x)) & (x not in intron_cols), identities.columns))\n",
    "identities['insertions'] = identities[insertion_cols].apply(lambda x: sum(x != '.'), axis=1)\n",
    "\n",
    "# Deletions at positions that are not the variable arm, and not counting 17/17a/20a/20b\n",
    "base_cols = list(filter(lambda x: bool(re.match('^\\d+$', x)) & (x not in ['74', '75', '76', '17', '17a', '20a', '20b']), identities.columns))\n",
    "identities['deletions'] = identities[base_cols].apply(lambda x: ''.join(x).count('-'), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loop sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def bounds_to_cols(cols, start, end):\n",
    "  selected_cols = []\n",
    "  for col in cols:\n",
    "    matches = re.findall('\\d+', col)\n",
    "    if len(matches) < 1: continue\n",
    "    index = int(matches[0])\n",
    "    if (index >= start and index <= end or col[0:3] == '{}i'.format(start - 1)) and col[0] != 'V':\n",
    "      selected_cols.append(col)\n",
    "  return selected_cols\n",
    "\n",
    "dloop_cols = list(filter(lambda col: ':' not in col, bounds_to_cols(identities.columns, 14, 21)))\n",
    "identities['D-loop'] = identities[dloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "# Leu, Ser have a 3 bp D stem\n",
    "dloop_II_cols = list(filter(lambda col: ':' not in col, bounds_to_cols(identities.columns, 13, 22)))\n",
    "identities.ix[(identities.isotype == 'Leu') | (identities.isotype == 'Ser'), 'D-loop'] = identities[dloop_II_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "acloop_cols = list(filter(lambda x: not re.match('37i.+', x), bounds_to_cols(identities.columns, 32, 38)))\n",
    "identities['AC-loop'] = identities[acloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "tpcloop_cols = bounds_to_cols(identities.columns, 54, 60)\n",
    "identities['TPC-loop'] = identities[tpcloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "varm_cols = list(filter(lambda x: ('V' in x) & (':' not in x), identities.columns))\n",
    "identities['V-arm'] = identities[varm_cols].apply(lambda x: len(x[(x != '.') & (x != '-') & (x != \"-:-\")]), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### High-quality tRNA set\n",
    "\n",
    "There was talk about using the high quality tRNAs to look at identity elements. Instead, I will just annotate specific tRNAs as high-quality or not. We can sort out the differences later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Import list of quality tRNAs\n",
    "quality_trnas = [line.strip() for line in open('quality-set.out')]\n",
    "identities['quality'] = identities.seqname.isin(quality_trnas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Restrict tRNAs by species\n",
    "\n",
    "We may also want to limit the contribution of any single species, similar to how we built the isotype-specific models. The restricted set is a subset of the quality set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "identities.ix[:, 'restrict'] = False\n",
    "for species in identities.species.unique():\n",
    "  species_df = identities.ix[identities.species == species, :]\n",
    "  for isotype in isotypes:\n",
    "    isotype_indices = identities.ix[(identities.species == species) & (identities.isotype == isotype) & (-identities.restrict), ].index\n",
    "    quality_indices = identities.ix[(identities.species == species) & (identities.isotype == isotype) & (-identities.restrict) & (identities.quality), ].index\n",
    "    unique_scores_indices = isotype_indices[-identities.ix[isotype_indices, 'score'].duplicated()]\n",
    "    unique_quality_indices = quality_indices[-identities.ix[quality_indices, 'score'].duplicated()]\n",
    "    \n",
    "    if len(unique_quality_indices) >= 50:\n",
    "      unique_quality_indices = unique_quality_indices[numpy.argsort(identities.ix[unique_quality_indices, 'score'])][::-1][:50]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_quality_indices))\n",
    "    elif len(unique_quality_indices) > 0:\n",
    "      if len(unique_scores_indices) > 50 - len(unique_quality_indices):\n",
    "        unique_scores_indices = unique_scores_indices[numpy.argsort(identities.ix[unique_scores_indices, 'score'])][::-1][:50 - len(unique_quality_indices)]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_quality_indices) - set(unique_scores_indices))\n",
    "    else:\n",
    "      if len(unique_scores_indices) > 50:\n",
    "        unique_scores_indices = unique_scores_indices[numpy.argsort(identities.ix[unique_scores_indices, 'score'])][::-1][:50]\n",
    "      restricted_indices = list(set(isotype_indices) - set(unique_scores_indices))\n",
    "    \n",
    "    identities.ix[restricted_indices, 'restrict'] = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The quality set does have some issues, though: it filters out rare codons, when rare is relative to model species. It also performs a hard check for anticodon-isotype model match. Fungi have much more diverged tRNAs and, in many cases, a different isotype scores higher even when it's clear that it's the bona fide tRNA (mostly based on number of tRNAs of that isotype, or codon usage). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export to R\n",
    "\n",
    "R has superior visualization capabilities.\n",
    "\n",
    "### Order columns\n",
    "\n",
    "To make it look pretty."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def position_str_to_int(position):\n",
    "  if position == \"20a\": return 20.1\n",
    "  if position == \"20b\": return 20.2\n",
    "  digits = re.findall('\\d+', position)\n",
    "  if len(digits) == 0: return -1\n",
    "  insert = 0\n",
    "  if 'i' in position and len(digits) == 2: insert = float(digits[1]) / 1000\n",
    "  if position[0] == 'V':\n",
    "    if ':' in position: return int(digits[0]) + 45 - 10 + insert # V11~V17\n",
    "    else: return int(digits[0]) + 45 + 7 + insert # V1~V5\n",
    "  if int(digits[0]) >= 46: return int(digits[0]) + 50 + insert # just add an arbitrarily large number to skip v-arm\n",
    "  return int(digits[0]) + insert\n",
    "\n",
    "identities = identities[sorted(list(identities.columns), key=position_str_to_int)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "identities.to_csv(path_or_buf='identities.tsv', sep='\\t', index_label=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clade</th>\n",
       "      <th>domain</th>\n",
       "      <th>isotype</th>\n",
       "      <th>seqname</th>\n",
       "      <th>species</th>\n",
       "      <th>species_long</th>\n",
       "      <th>taxid</th>\n",
       "      <th>isotype_best</th>\n",
       "      <th>anticodon</th>\n",
       "      <th>score</th>\n",
       "      <th>isoscore</th>\n",
       "      <th>isoscore_ac</th>\n",
       "      <th>GC</th>\n",
       "      <th>intron</th>\n",
       "      <th>insertions</th>\n",
       "      <th>deletions</th>\n",
       "      <th>D-loop</th>\n",
       "      <th>AC-loop</th>\n",
       "      <th>TPC-loop</th>\n",
       "      <th>V-arm</th>\n",
       "      <th>quality</th>\n",
       "      <th>restrict</th>\n",
       "      <th>1:72</th>\n",
       "      <th>1</th>\n",
       "      <th>1i1</th>\n",
       "      <th>2:71</th>\n",
       "      <th>2</th>\n",
       "      <th>2i1</th>\n",
       "      <th>3:70</th>\n",
       "      <th>3</th>\n",
       "      <th>3i1</th>\n",
       "      <th>3i2</th>\n",
       "      <th>3i3</th>\n",
       "      <th>3i4</th>\n",
       "      <th>3i5</th>\n",
       "      <th>3i6</th>\n",
       "      <th>4:69</th>\n",
       "      <th>4</th>\n",
       "      <th>4i1</th>\n",
       "      <th>4i2</th>\n",
       "      <th>4i3</th>\n",
       "      <th>4i4</th>\n",
       "      <th>4i5</th>\n",
       "      <th>4i6</th>\n",
       "      <th>4i7</th>\n",
       "      <th>4i8</th>\n",
       "      <th>4i9</th>\n",
       "      <th>4i10</th>\n",
       "      <th>4i11</th>\n",
       "      <th>5:68</th>\n",
       "      <th>5</th>\n",
       "      <th>5i1</th>\n",
       "      <th>5i2</th>\n",
       "      <th>5i3</th>\n",
       "      <th>5i4</th>\n",
       "      <th>5i5</th>\n",
       "      <th>5i6</th>\n",
       "      <th>6:67</th>\n",
       "      <th>6</th>\n",
       "      <th>6i1</th>\n",
       "      <th>7:66</th>\n",
       "      <th>7</th>\n",
       "      <th>7i1</th>\n",
       "      <th>7i2</th>\n",
       "      <th>7i3</th>\n",
       "      <th>7i4</th>\n",
       "      <th>7i5</th>\n",
       "      <th>7i6</th>\n",
       "      <th>7i7</th>\n",
       "      <th>7i8</th>\n",
       "      <th>7i9</th>\n",
       "      <th>8</th>\n",
       "      <th>8:14:21</th>\n",
       "      <th>8:14</th>\n",
       "      <th>8i1</th>\n",
       "      <th>8i2</th>\n",
       "      <th>9</th>\n",
       "      <th>9:12:23</th>\n",
       "      <th>9:23</th>\n",
       "      <th>9i1</th>\n",
       "      <th>10:25</th>\n",
       "      <th>10</th>\n",
       "      <th>10:25:45</th>\n",
       "      <th>10:45</th>\n",
       "      <th>10i1</th>\n",
       "      <th>11:24</th>\n",
       "      <th>11</th>\n",
       "      <th>12:23</th>\n",
       "      <th>12</th>\n",
       "      <th>12i1</th>\n",
       "      <th>13:22</th>\n",
       "      <th>13</th>\n",
       "      <th>13:22:46</th>\n",
       "      <th>14</th>\n",
       "      <th>14i1</th>\n",
       "      <th>14i2</th>\n",
       "      <th>14i3</th>\n",
       "      <th>14i4</th>\n",
       "      <th>14i5</th>\n",
       "      <th>14i6</th>\n",
       "      <th>14i7</th>\n",
       "      <th>14i8</th>\n",
       "      <th>14i9</th>\n",
       "      <th>14i10</th>\n",
       "      <th>14i11</th>\n",
       "      <th>14i12</th>\n",
       "      <th>14i13</th>\n",
       "      <th>15</th>\n",
       "      <th>15:48</th>\n",
       "      <th>16</th>\n",
       "      <th>16i1</th>\n",
       "      <th>16i2</th>\n",
       "      <th>16i3</th>\n",
       "      <th>16i4</th>\n",
       "      <th>16i5</th>\n",
       "      <th>16i6</th>\n",
       "      <th>16i7</th>\n",
       "      <th>16i8</th>\n",
       "      <th>16i9</th>\n",
       "      <th>16i10</th>\n",
       "      <th>16i11</th>\n",
       "      <th>16i12</th>\n",
       "      <th>16i13</th>\n",
       "      <th>16i14</th>\n",
       "      <th>16i15</th>\n",
       "      <th>16i16</th>\n",
       "      <th>17</th>\n",
       "      <th>17a</th>\n",
       "      <th>18</th>\n",
       "      <th>18:55</th>\n",
       "      <th>19</th>\n",
       "      <th>19:56</th>\n",
       "      <th>19i1</th>\n",
       "      <th>19i2</th>\n",
       "      <th>19i3</th>\n",
       "      <th>19i4</th>\n",
       "      <th>19i5</th>\n",
       "      <th>19i6</th>\n",
       "      <th>19i7</th>\n",
       "      <th>19i8</th>\n",
       "      <th>19i9</th>\n",
       "      <th>19i10</th>\n",
       "      <th>20</th>\n",
       "      <th>20i1</th>\n",
       "      <th>20i2</th>\n",
       "      <th>20i3</th>\n",
       "      <th>20i4</th>\n",
       "      <th>20a</th>\n",
       "      <th>20b</th>\n",
       "      <th>21</th>\n",
       "      <th>22</th>\n",
       "      <th>22:46</th>\n",
       "      <th>22i1</th>\n",
       "      <th>23</th>\n",
       "      <th>23i1</th>\n",
       "      <th>23i2</th>\n",
       "      <th>23i3</th>\n",
       "      <th>23i4</th>\n",
       "      <th>23i5</th>\n",
       "      <th>23i6</th>\n",
       "      <th>24</th>\n",
       "      <th>24i1</th>\n",
       "      <th>25</th>\n",
       "      <th>25i1</th>\n",
       "      <th>25i2</th>\n",
       "      <th>25i3</th>\n",
       "      <th>25i4</th>\n",
       "      <th>26</th>\n",
       "      <th>26:44</th>\n",
       "      <th>26i1</th>\n",
       "      <th>26i2</th>\n",
       "      <th>26i3</th>\n",
       "      <th>26i4</th>\n",
       "      <th>26i5</th>\n",
       "      <th>26i6</th>\n",
       "      <th>26i7</th>\n",
       "      <th>26i8</th>\n",
       "      <th>26i9</th>\n",
       "      <th>26i10</th>\n",
       "      <th>27:43</th>\n",
       "      <th>27</th>\n",
       "      <th>27i1</th>\n",
       "      <th>27i2</th>\n",
       "      <th>27i3</th>\n",
       "      <th>27i4</th>\n",
       "      <th>28:42</th>\n",
       "      <th>28</th>\n",
       "      <th>28i1</th>\n",
       "      <th>29:41</th>\n",
       "      <th>29</th>\n",
       "      <th>29i1</th>\n",
       "      <th>29i2</th>\n",
       "      <th>30:40</th>\n",
       "      <th>30</th>\n",
       "      <th>30i1</th>\n",
       "      <th>31:39</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>35i1</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>37i1</th>\n",
       "      <th>37i2</th>\n",
       "      <th>37i3</th>\n",
       "      <th>37i4</th>\n",
       "      <th>37i5</th>\n",
       "      <th>37i6</th>\n",
       "      <th>37i7</th>\n",
       "      <th>37i8</th>\n",
       "      <th>37i9</th>\n",
       "      <th>38</th>\n",
       "      <th>38i1</th>\n",
       "      <th>39</th>\n",
       "      <th>39i1</th>\n",
       "      <th>40</th>\n",
       "      <th>40i1</th>\n",
       "      <th>40i2</th>\n",
       "      <th>41</th>\n",
       "      <th>41i1</th>\n",
       "      <th>41i2</th>\n",
       "      <th>42</th>\n",
       "      <th>42i1</th>\n",
       "      <th>43</th>\n",
       "      <th>44</th>\n",
       "      <th>44i1</th>\n",
       "      <th>44i2</th>\n",
       "      <th>44i3</th>\n",
       "      <th>44i4</th>\n",
       "      <th>44i5</th>\n",
       "      <th>44i6</th>\n",
       "      <th>44i7</th>\n",
       "      <th>44i8</th>\n",
       "      <th>44i9</th>\n",
       "      <th>44i10</th>\n",
       "      <th>44i11</th>\n",
       "      <th>44i12</th>\n",
       "      <th>44i13</th>\n",
       "      <th>44i14</th>\n",
       "      <th>44i15</th>\n",
       "      <th>44i16</th>\n",
       "      <th>44i17</th>\n",
       "      <th>44i18</th>\n",
       "      <th>44i19</th>\n",
       "      <th>44i20</th>\n",
       "      <th>44i21</th>\n",
       "      <th>44i22</th>\n",
       "      <th>44i23</th>\n",
       "      <th>45</th>\n",
       "      <th>V11:V21</th>\n",
       "      <th>V12:V22</th>\n",
       "      <th>V13:V23</th>\n",
       "      <th>V14:V24</th>\n",
       "      <th>V15:V25</th>\n",
       "      <th>V16:V26</th>\n",
       "      <th>V17:V27</th>\n",
       "      <th>V1</th>\n",
       "      <th>V2</th>\n",
       "      <th>V3</th>\n",
       "      <th>V4</th>\n",
       "      <th>V5</th>\n",
       "      <th>V11</th>\n",
       "      <th>V12</th>\n",
       "      <th>V13</th>\n",
       "      <th>V14</th>\n",
       "      <th>V15</th>\n",
       "      <th>V16</th>\n",
       "      <th>V17</th>\n",
       "      <th>V21</th>\n",
       "      <th>V22</th>\n",
       "      <th>V23</th>\n",
       "      <th>V24</th>\n",
       "      <th>V25</th>\n",
       "      <th>V26</th>\n",
       "      <th>V27</th>\n",
       "      <th>46</th>\n",
       "      <th>47</th>\n",
       "      <th>47i1</th>\n",
       "      <th>47i2</th>\n",
       "      <th>47i3</th>\n",
       "      <th>48</th>\n",
       "      <th>49:65</th>\n",
       "      <th>49</th>\n",
       "      <th>49i1</th>\n",
       "      <th>50:64</th>\n",
       "      <th>50</th>\n",
       "      <th>50i1</th>\n",
       "      <th>50i2</th>\n",
       "      <th>50i3</th>\n",
       "      <th>50i4</th>\n",
       "      <th>50i5</th>\n",
       "      <th>50i6</th>\n",
       "      <th>50i7</th>\n",
       "      <th>51:63</th>\n",
       "      <th>51</th>\n",
       "      <th>51i1</th>\n",
       "      <th>51i2</th>\n",
       "      <th>51i3</th>\n",
       "      <th>52:62</th>\n",
       "      <th>52</th>\n",
       "      <th>52i1</th>\n",
       "      <th>53:61</th>\n",
       "      <th>53</th>\n",
       "      <th>53i1</th>\n",
       "      <th>54</th>\n",
       "      <th>54:58</th>\n",
       "      <th>54i1</th>\n",
       "      <th>54i2</th>\n",
       "      <th>54i3</th>\n",
       "      <th>54i4</th>\n",
       "      <th>55</th>\n",
       "      <th>55i1</th>\n",
       "      <th>55i2</th>\n",
       "      <th>55i3</th>\n",
       "      <th>55i4</th>\n",
       "      <th>55i5</th>\n",
       "      <th>56</th>\n",
       "      <th>56i1</th>\n",
       "      <th>56i2</th>\n",
       "      <th>56i3</th>\n",
       "      <th>56i4</th>\n",
       "      <th>57</th>\n",
       "      <th>57i1</th>\n",
       "      <th>58</th>\n",
       "      <th>58i1</th>\n",
       "      <th>58i2</th>\n",
       "      <th>59</th>\n",
       "      <th>59i1</th>\n",
       "      <th>60</th>\n",
       "      <th>60i1</th>\n",
       "      <th>60i2</th>\n",
       "      <th>60i3</th>\n",
       "      <th>60i4</th>\n",
       "      <th>60i5</th>\n",
       "      <th>60i6</th>\n",
       "      <th>60i7</th>\n",
       "      <th>60i8</th>\n",
       "      <th>60i9</th>\n",
       "      <th>60i10</th>\n",
       "      <th>60i11</th>\n",
       "      <th>61</th>\n",
       "      <th>61i1</th>\n",
       "      <th>62</th>\n",
       "      <th>63</th>\n",
       "      <th>64</th>\n",
       "      <th>64i1</th>\n",
       "      <th>64i2</th>\n",
       "      <th>65</th>\n",
       "      <th>65i1</th>\n",
       "      <th>65i2</th>\n",
       "      <th>65i3</th>\n",
       "      <th>65i4</th>\n",
       "      <th>65i5</th>\n",
       "      <th>65i6</th>\n",
       "      <th>65i7</th>\n",
       "      <th>65i8</th>\n",
       "      <th>65i9</th>\n",
       "      <th>66</th>\n",
       "      <th>66i1</th>\n",
       "      <th>67</th>\n",
       "      <th>67i1</th>\n",
       "      <th>68</th>\n",
       "      <th>68i1</th>\n",
       "      <th>68i2</th>\n",
       "      <th>69</th>\n",
       "      <th>69i1</th>\n",
       "      <th>70</th>\n",
       "      <th>70i1</th>\n",
       "      <th>70i2</th>\n",
       "      <th>70i3</th>\n",
       "      <th>70i4</th>\n",
       "      <th>70i5</th>\n",
       "      <th>70i6</th>\n",
       "      <th>70i7</th>\n",
       "      <th>70i8</th>\n",
       "      <th>70i9</th>\n",
       "      <th>71</th>\n",
       "      <th>71i1</th>\n",
       "      <th>71i2</th>\n",
       "      <th>72</th>\n",
       "      <th>73</th>\n",
       "      <th>74</th>\n",
       "      <th>75</th>\n",
       "      <th>76</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mammalia</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Leu</td>\n",
       "      <td>ailMel1_GL192338.1.trna142-LeuCAG</td>\n",
       "      <td>ailMel1</td>\n",
       "      <td>Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)</td>\n",
       "      <td>.</td>\n",
       "      <td>Leu</td>\n",
       "      <td>CAG</td>\n",
       "      <td>77.3</td>\n",
       "      <td>124.5</td>\n",
       "      <td>124.5</td>\n",
       "      <td>0.463415</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:G</td>\n",
       "      <td>G:G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:-</td>\n",
       "      <td>G:-</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>C:G</td>\n",
       "      <td>U:A</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C:G</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mammalia</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Leu</td>\n",
       "      <td>ailMel1_GL192338.1.trna384-LeuCAG</td>\n",
       "      <td>ailMel1</td>\n",
       "      <td>Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)</td>\n",
       "      <td>.</td>\n",
       "      <td>Leu</td>\n",
       "      <td>CAG</td>\n",
       "      <td>78.2</td>\n",
       "      <td>124.9</td>\n",
       "      <td>124.9</td>\n",
       "      <td>0.487805</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:G</td>\n",
       "      <td>G:G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:-</td>\n",
       "      <td>G:-</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:G</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>C:G</td>\n",
       "      <td>U:A</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C:G</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mammalia</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Thr</td>\n",
       "      <td>ailMel1_GL192339.1.trna304-ThrCGT</td>\n",
       "      <td>ailMel1</td>\n",
       "      <td>Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)</td>\n",
       "      <td>.</td>\n",
       "      <td>Thr</td>\n",
       "      <td>CGT</td>\n",
       "      <td>78.3</td>\n",
       "      <td>93.8</td>\n",
       "      <td>93.8</td>\n",
       "      <td>0.439024</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:G</td>\n",
       "      <td>G:G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:G</td>\n",
       "      <td>G:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>A:A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:A:A</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:A</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mammalia</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Ile</td>\n",
       "      <td>ailMel1_GL192340.1.trna29-IleTAT</td>\n",
       "      <td>ailMel1</td>\n",
       "      <td>Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)</td>\n",
       "      <td>.</td>\n",
       "      <td>Ile</td>\n",
       "      <td>TAT</td>\n",
       "      <td>73.2</td>\n",
       "      <td>105.7</td>\n",
       "      <td>105.7</td>\n",
       "      <td>0.414634</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:G:C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:G</td>\n",
       "      <td>G:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>C:G:G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A:U</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mammalia</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Arg</td>\n",
       "      <td>ailMel1_GL192348.1.trna445-ArgACG</td>\n",
       "      <td>ailMel1</td>\n",
       "      <td>Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)</td>\n",
       "      <td>.</td>\n",
       "      <td>Arg</td>\n",
       "      <td>ACG</td>\n",
       "      <td>64.2</td>\n",
       "      <td>98.9</td>\n",
       "      <td>98.9</td>\n",
       "      <td>0.365854</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A:A</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:G:C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C:G</td>\n",
       "      <td>G:G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>C:A:A</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A:U</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>G:U</td>\n",
       "      <td>G</td>\n",
       "      <td>G:C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A:A</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-:-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>C:G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U:G</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>A:U</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>G:C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>U:A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>U</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>U</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      clade     domain isotype                            seqname  species                                                       species_long taxid isotype_best anticodon  score  isoscore  isoscore_ac        GC  intron  insertions  deletions  D-loop  AC-loop  TPC-loop  V-arm quality restrict 1:72  1 1i1 2:71  2 2i1 3:70  3 3i1 3i2 3i3 3i4 3i5 3i6 4:69  4 4i1 4i2 4i3 4i4 4i5 4i6 4i7 4i8 4i9 4i10 4i11 5:68  5 5i1 5i2 5i3 5i4 5i5 5i6 6:67  6 6i1 7:66  7 7i1 7i2 7i3 7i4 7i5 7i6 7i7 7i8 7i9  8 8:14:21 8:14 8i1 8i2  9 9:12:23 9:23 9i1 10:25 10 10:25:45 10:45 10i1 11:24 11 12:23 12 12i1 13:22 13 13:22:46 14 14i1 14i2 14i3 14i4 14i5 14i6 14i7 14i8 14i9 14i10 14i11 14i12 14i13 15 15:48 16 16i1 16i2 16i3 16i4 16i5 16i6 16i7 16i8 16i9 16i10 16i11 16i12 16i13 16i14 16i15 16i16 17 17a 18 18:55 19 19:56 19i1 19i2 19i3 19i4 19i5 19i6 19i7 19i8 19i9 19i10 20 20i1 20i2 20i3 20i4 20a 20b 21 22 22:46 22i1 23 23i1 23i2 23i3 23i4 23i5 23i6 24 24i1 25 25i1 25i2 25i3 25i4 26 26:44 26i1 26i2 26i3 26i4 26i5 26i6 26i7 26i8 26i9 26i10 27:43 27 27i1 27i2 27i3 27i4 28:42 28 28i1 29:41 29 29i1 29i2 30:40 30 30i1 31:39 31 32 33 34 35 35i1 36 37 37i1 37i2 37i3 37i4 37i5 37i6 37i7 37i8 37i9 38 38i1 39 39i1 40 40i1 40i2 41 41i1 41i2 42 42i1 43 44 44i1 44i2 44i3 44i4 44i5 44i6 44i7 44i8 44i9 44i10 44i11 44i12 44i13 44i14 44i15 44i16 44i17 44i18 44i19 44i20 44i21 44i22 44i23 45 V11:V21 V12:V22 V13:V23 V14:V24 V15:V25 V16:V26 V17:V27 V1 V2 V3 V4 V5 V11 V12 V13 V14 V15 V16 V17 V21 V22 V23 V24 V25 V26 V27 46 47 47i1 47i2 47i3 48 49:65 49 49i1 50:64 50 50i1 50i2 50i3 50i4 50i5 50i6 50i7 51:63 51 51i1 51i2 51i3 52:62 52 52i1 53:61 53 53i1 54 54:58 54i1 54i2 54i3 54i4 55 55i1 55i2 55i3 55i4 55i5 56 56i1 56i2 56i3 56i4 57 57i1 58 58i1 58i2 59 59i1 60 60i1 60i2 60i3 60i4 60i5 60i6 60i7 60i8 60i9 60i10 60i11 61 61i1 62 63 64 64i1 64i2 65 65i1 65i2 65i3 65i4 65i5 65i6 65i7 65i8 65i9 66 66i1 67 67i1 68 68i1 68i2 69 69i1 70 70i1 70i2 70i3 70i4 70i5 70i6 70i7 70i8 70i9 71 71i1 71i2 72 73 74 75 76\n",
       "0  Mammalia  eukaryota     Leu  ailMel1_GL192338.1.trna142-LeuCAG  ailMel1  Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)     .          Leu       CAG   77.3     124.5        124.5  0.463415       0           0          2      11        7         8     11    True    False  G:C  G   .  U:A  U   .  C:G  C   .   .   .   .   .   .  A:U  A   .   .   .   .   .   .   .   .   .    .    .  G:C  G   .   .   .   .   .   .  G:U  G   .  A:U  A   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:C:G  G:G   .   G:C  G    G:C:-   G:-    .   C:G  C   C:G  C    .   G:A  G    G:A:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  C    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   C   U  A  A   A:G    .  G    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:U    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   U:A  U    .   G:C  G    .    .   C:G  C    .   G:C  G  U  U  C  A    .  G  G    .    .    .    .    .    .    .    .    .  U    .  C    .  G    .    .  C    .    .  A    .  G  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  -     C:G     U:A     C:G     C:G     -:-     -:-     -:-  C  C  U  -  -   C   U   C   C   -   -   -   G   A   G   G   -   -   -  G  -    .    .    .  C   G:C  G    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   G:C  G    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  A    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  C  C  A    .    .  C    .    .    .    .    .    .    .    .    .  U    .  U    .  C    .    .  U    .  G    .    .    .    .    .    .    .    .    .  A    .    .  C  A  -  -  -\n",
       "1  Mammalia  eukaryota     Leu  ailMel1_GL192338.1.trna384-LeuCAG  ailMel1  Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)     .          Leu       CAG   78.2     124.9        124.9  0.487805       0           0          2      11        7         8     11    True    False  G:C  G   .  U:A  U   .  C:G  C   .   .   .   .   .   .  A:U  A   .   .   .   .   .   .   .   .   .    .    .  G:C  G   .   .   .   .   .   .  G:C  G   .  A:U  A   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:C:G  G:G   .   G:C  G    G:C:-   G:-    .   C:G  C   C:G  C    .   G:A  G    G:A:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  C    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   C   U  A  A   A:G    .  G    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:U    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   U:A  U    .   G:C  G    .    .   C:G  C    .   G:C  G  U  U  C  A    .  G  G    .    .    .    .    .    .    .    .    .  U    .  C    .  G    .    .  C    .    .  A    .  G  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  -     C:G     U:A     C:G     C:G     -:-     -:-     -:-  C  C  U  -  -   C   U   C   C   -   -   -   G   A   G   G   -   -   -  G  -    .    .    .  C   G:C  G    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   G:C  G    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  A    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  C  C  A    .    .  C    .    .    .    .    .    .    .    .    .  U    .  C    .  C    .    .  U    .  G    .    .    .    .    .    .    .    .    .  A    .    .  C  A  -  -  -\n",
       "2  Mammalia  eukaryota     Thr  ailMel1_GL192339.1.trna304-ThrCGT  ailMel1  Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)     .          Thr       CGT   78.3      93.8         93.8  0.439024       0           0          0       7        7         8      0    True    False  G:C  G   .  G:C  G   .  C:G  C   .   .   .   .   .   .  G:U  G   .   .   .   .   .   .   .   .   .    .    .  C:G  C   .   .   .   .   .   .  G:C  G   .  G:C  G   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:C:G  G:G   .   G:C  G    G:C:G   G:G    .   C:G  C   C:G  C    .   A:A  A    A:A:A  A    .    .    .    .    .    .    .    .    .     .     .     .     .  G   G:C  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   -   -  A  A   A:A    .  G    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   U:A  U    .    .    .    .   C:G  C    .   G:C  G    .    .   G:C  G    .   U:A  U  C  U  C  G    .  U  A    .    .    .    .    .    .    .    .    .  A    .  A    .  C    .    .  C    .    .  G    .  A  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  G     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  A  U    .    .    .  C   G:U  G    .   C:G  C    .    .    .    .    .    .    .   G:C  G    .    .    .   G:C  G    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  A    .  C    .    .    .    .    .    .    .    .    .     .     .  C    .  C  C  G    .    .  U    .    .    .    .    .    .    .    .    .  C    .  C    .  G    .    .  U    .  G    .    .    .    .    .    .    .    .    .  C    .    .  C  U  -  -  -\n",
       "3  Mammalia  eukaryota     Ile   ailMel1_GL192340.1.trna29-IleTAT  ailMel1  Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)     .          Ile       TAT   73.2     105.7        105.7  0.414634       0           0          0       9        7         8      0    True    False  G:C  G   .  C:G  C   .  U:A  U   .   .   .   .   .   .  C:G  C   .   .   .   .   .   .   .   .   .    .    .  C:G  C   .   .   .   .   .   .  A:U  A   .  G:C  G   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:G:C  G:C   .   G:C  G    G:C:G   G:G    .   C:G  C   G:C  G    .   C:G  C    C:G:G  A    .    .    .    .    .    .    .    .    .     .     .     .     .  A   A:U  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  C   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  U    .    .    .    .   U   -  A  G   G:G    .  C    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   C:G  C    .    .    .    .   G:C  G    .   G:C  G    .    .   U:G  U    .   A:U  A  C  U  U  A    .  U  A    .    .    .    .    .    .    .    .    .  A    .  U    .  G    .    .  C    .    .  C    .  G  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  G     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  G  U    .    .    .  U   G:C  G    .   U:A  U    .    .    .    .    .    .    .   G:C  G    .    .    .   A:U  A    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  G    .  C    .    .    .    .    .    .    .    .    .     .     .  C    .  U  C  A    .    .  C    .    .    .    .    .    .    .    .    .  C    .  U    .  G    .    .  G    .  A    .    .    .    .    .    .    .    .    .  G    .    .  C  A  -  -  -\n",
       "4  Mammalia  eukaryota     Arg  ailMel1_GL192348.1.trna445-ArgACG  ailMel1  Ailuropoda melanoleuca (panda) (BGI-Shenzhen AilMel 1.0 Dec 2009)     .          Arg       ACG   64.2      98.9         98.9  0.365854       0           0          0       8        7         8      0    True    False  G:C  G   .  G:U  G   .  G:C  G   .   .   .   .   .   .  C:G  C   .   .   .   .   .   .   .   .   .    .    .  C:G  C   .   .   .   .   .   .  A:U  A   .  G:C  G   .   .   .   .   .   .   .   .   .  U   U:A:A  U:A   .   .  G   G:G:C  G:C   .   G:C  G    G:C:G   G:G    .   C:G  C   G:C  G    .   C:A  C    C:A:A  A    .    .    .    .    .    .    .    .    .     .     .     .     .  A   A:U  U    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .  -   -  G   G:U  G   G:C    .    .    .    .    .    .    .    .    .     .  A    .    .    .    .   U   -  A  A   A:A    .  C    .    .    .    .    .    .  G    .  C    .    .    .    .  G   G:A    .    .    .    .    .    .    .    .    .     .   U:A  U    .    .    .    .   C:G  C    .   U:A  U    .    .   G:C  G    .   A:U  A  C  U  A  C    .  G  G    .    .    .    .    .    .    .    .    .  A    .  U    .  C    .    .  A    .    .  G    .  A  A    .    .    .    .    .    .    .    .    .     .     .     .     .     .     .     .     .     .     .     .     .     .     .  G     -:-     -:-     -:-     -:-     -:-     -:-     -:-  -  -  -  -  -   -   -   -   -   -   -   -   -   -   -   -   -   -   -  A  U    .    .    .  U   C:G  C    .   U:G  U    .    .    .    .    .    .    .   A:U  A    .    .    .   G:C  G    .   G:C  G    .  U   U:A    .    .    .    .  U    .    .    .    .    .  C    .    .    .    .  G    .  A    .    .  C    .  U    .    .    .    .    .    .    .    .    .     .     .  C    .  C  U  G    .    .  G    .    .    .    .    .    .    .    .    .  C    .  U    .  G    .    .  G    .  C    .    .    .    .    .    .    .    .    .  U    .    .  C  G  -  -  -"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "identities.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Resolve consensus positions\n",
    "\n",
    "There are some discrepencies between the covariance model, the Sprinzl model, and reality. For instance, is position 17 really a gap position for most tRNAs? If so, why is it considered a position? We know that the D loop interacts with the TPC loop, so the general position 17 area should be a key identity element or structural feature contributing to tRNA activity. What should we call insertions at 17 if it's not a real position? These conflicts need to be resolved into a consensus version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def bounds_to_cols(cols, start, end):\n",
    "  selected_cols = []\n",
    "  for col in cols:\n",
    "    matches = re.findall('\\d+', col)\n",
    "    if len(matches) < 1: continue\n",
    "    index = int(matches[0])\n",
    "    if (index >= start and index <= end or col[0:3] == '{}i'.format(start - 1)) and col[0] != 'V':\n",
    "      selected_cols.append(col)\n",
    "  return selected_cols\n",
    "\n",
    "def position_histogram(identities, cols):\n",
    "  df = pd.melt(identities[cols + ['seqname']], id_vars=['seqname'])\n",
    "  df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "  counts = pd.Series(Counter([x for x in df.variable]), index=cols)\n",
    "  counts = counts[counts > 10]\n",
    "  counts = pd.DataFrame({\"Position\": counts.index, 'tRNAs': counts})\n",
    "  return counts.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### D-loop positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dloop_cols = bounds_to_cols(identities.columns, 14, 21)\n",
    "position_histogram(identities, dloop_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the model curated using --hand, the D loop now contains 17a, 20a, and 20b. The alignment seems to have been fixed, as we don't have a bump at 20i5 (or even 20i5-20i7) anymore.\n",
    "\n",
    "#### 17A\n",
    "\n",
    "What isotypes have position 17?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(identities[['isotype', '17']], id_vars=['isotype'])\n",
    "df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "df = pd.Series(Counter([x for x in df.isotype]))\n",
    "df.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(identities[['isotype', '17a']], id_vars=['isotype'])\n",
    "df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "df = pd.Series(Counter([x for x in df.isotype]))\n",
    "df.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on Sprinzl alignments, there are 19 tRNAs out of 2249 with a base at position 17A. It also doesn't look like we can hardcode position 17 for just one or two isotypes - it's universal enough. Based on this data, 17A is pretty rare, and even position 17 is only there 30% of the time (based on our data). Position 17A could actually be considered as an insertion.\n",
    "\n",
    "#### 20a/20b\n",
    "\n",
    "** This section is somewhat deprecated as the alignment has been fixed. ** The following analysis has been modified to double check whether the new 20a/20b alignment is sound.\n",
    "\n",
    "Also based on Sprinzl alignments, no tRNAs have only 20b and not 20a. So whenever there is 20b, 20a is sure to be there. Let's take a look at the distribution out of 2249 tRNA genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sprinzl_20ab = pd.read_table('sprinzl-20ab.txt', header=None, names=['20a', '20b', 'Isotype'])\n",
    "sprinzl_20ab = sprinzl_20ab.groupby([\"Isotype\"]).aggregate(lambda x: len(x[x != '-']))\n",
    "sprinzl_20ab.plot(kind=\"bar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This complicates things. Ideally we'd like to see all of the tRNAs with 20b to belong to one or two isotypes. Then we could manually check the most common insertion positions for those isotypes and adjust it. Instead, we'll just have to check the insertion positions for all isotypes and attempt to come up with some universal rules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sprinzl_20ab = pd.read_table('sprinzl-20ab.txt', header=None, names=['20a', '20b', 'Isotype'])\n",
    "sprinzl_20ab = sprinzl_20ab.groupby([\"20a\", \"20b\"]).aggregate(len)\n",
    "sprinzl_20ab.plot(kind=\"bar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This plot tells us a few things.\n",
    "\n",
    "First, 20a is typically a T, but there are a few ACG 20as. To be expected, as 20ab are typically modified to dihydrouridine (D).\n",
    "\n",
    "20b is typically also a T, but is most commonly paired with C20a. There are a few of the others, too. [Nissan and Perona (2000)](https://www.ncbi.nlm.nih.gov/pubmed/11105758) corroborate this by showing a successful replacement of  yeast function serine U20a/G20b with C20a/A20b _in vitro_. This complicates things. \n",
    "\n",
    "I hesitate to use a rule like \"First base is 20a, second base is 20b, third and up is 20i\". There is high potential for misannotation of an insertion as 20ab or vice versa. We can agree on the following:\n",
    "- single base insertion: 20a\n",
    "- two base insertion: 20a, 20b\n",
    "\n",
    "For three and up, we could potentially leave it unannotated. Let's take a look at the relative insertion lengths, by isotype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "insert_df = identities[['isotype'] + ['20a', '20b', '20i1', '20i2', '20i3', '20i4']]\n",
    "insertionab = insert_df.ix[:, 1:].apply(lambda x: sum((x != '.') & (x != '-')), axis=1)\n",
    "insert_df = insert_df.assign(insertionab = insertionab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(insert_df.groupby('insertionab').aggregate({\"isotype\": len}))\n",
    "plot = sns.violinplot(data=insert_df, x='isotype', y='insertionab', scale='width', cut=0)\n",
    "plot.set(ylim=(-1, 5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Half of the tRNAs have no 20a/20b, and the other half does. For 10 isotypes, all tRNAs are 2 insertions and below. For the other 10, it varies. Exactly how many tRNAs are we working with here? Is there a correlation betweeen number of tRNAs and number of tRNAs with insertions > 2?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df = insert_df.ix[insert_df.insertionab > 2, ['isotype', 'insertionab']].groupby('isotype').aggregate(len)\n",
    "print(\"No. tRNAs with > 2 insertions: {}\".format(sum(df.insertionab)))\n",
    "df = df.merge(identities[['isotype', 'seqname']].groupby('isotype').aggregate(len), how='right', right_index=True, left_index=True)\n",
    "df = df.fillna(0)\n",
    "df.columns = ['insertionab_tRNAs', 'tRNAs']\n",
    "\n",
    "# scatter plot\n",
    "plot = df.plot(kind='scatter', x='insertionab_tRNAs', y='tRNAs')\n",
    "for isotype, num_tRNAs in df.iterrows(): plot.annotate(isotype, num_tRNAs, xytext=(5, 0), textcoords='offset points', family='sans-serif')\n",
    "\n",
    "# add trendline\n",
    "beta = numpy.polyfit(x=df.insertionab_tRNAs, y=df.tRNAs, deg=1)\n",
    "trendline = pd.DataFrame({'x': list(df.insertionab_tRNAs) + [-20, 120], 'trendline': numpy.poly1d(beta)(list(df.insertionab_tRNAs) + [-20, 120])})\n",
    "trendline.plot(ax=plot, x='x', y='trendline', kind='line', xlim=[-3, 10], ylim=[0, 2500])\n",
    "\n",
    "plot.set_xlabel('No. tRNAs with >2 insertions at position 20')\n",
    "plot.set_ylabel('No. tRNAs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There's only 38 tRNAs with a 3 base or larger insertion. Here they are below. They're mostly leucine tRNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = insert_df[insert_df.insertionab > 2]\n",
    "df[\"seq\"] = df[['20a', '20b', '20i1', '20i2', '20i3', '20i4']].apply(lambda x: ''.join(x[(x != '.') & (x != '-')]), axis=1)\n",
    "df.sort_values('insertionab')[['isotype', 'seq']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variable stem-loop plus spacer nucleotides\n",
    "\n",
    "** This section is somewhat deprecated as the alignment has been fixed. ** The following analysis has been modified to double check whether the position 47 and V-arm alignment is sound."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "varm_cols = list(filter(lambda x: ('V' in x) & (':' not in x), identities.columns))\n",
    "varm_cols += bounds_to_cols(identities.columns, 44, 48)\n",
    "position_histogram(identities, varm_cols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data correlates well with the Sprinzl data. Most variable stems are missing at least V15:V25, V16:V26, and V17:V27, and most variable loops are 3 nucleotides long.\n",
    "\n",
    "Here is the same plot, but split into Ser, Leu, and Tyr, versus everything else."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(identities[varm_cols + ['seqname', 'isotype']], id_vars=['seqname', 'isotype'])\n",
    "df = df[(df.value != \".\") & (df.value != \"-\")]\n",
    "counts = pd.Series(Counter([x for x in df[(df.isotype == \"Tyr\") | (df.isotype == \"Leu\") | (df.isotype == \"Ser\")].variable]), index=varm_cols)\n",
    "counts = pd.DataFrame({\"Position\": counts.index, 'tRNAs': counts})\n",
    "counts2 = pd.Series(Counter([x for x in df[(df.isotype != \"Tyr\") | (df.isotype != \"Leu\") | (df.isotype != \"Ser\")].variable]), index=varm_cols)\n",
    "counts = pd.merge(counts, pd.DataFrame({\"Position\": counts2.index, 'tRNAs': counts2}), on=['Position'])\n",
    "counts.columns = ['Position', 'Class II', 'Class I']\n",
    "counts.index = counts.Position\n",
    "counts.plot(kind='bar')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to apply a similar analysis for 47 as we did for 20ab. What does the distribution of 47 + insert length look like?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "insert_df = identities[['isotype', '47', '47i1']]\n",
    "insert_df.head()\n",
    "insertion47 = 2 - sum([(insert_df.ix[:, 1] == \".\"), (insert_df.ix[:, 2] == \".\"), (insert_df.ix[:, 1] == \"-\"), (insert_df.ix[:, 2] == \"-\")])\n",
    "insert_df = insert_df.assign(insertion47 = insertion47)\n",
    "print(insert_df.groupby('insertion47').aggregate({\"isotype\": len}))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of the 94197 tRNAs with 1 nucleotide at position 47 + 47ix, which position does it appear on? Is there a typical position 47 nucleotide?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pd.melt(insert_df.ix[insert_df.insertion47 == 1, 0:3], id_vars='isotype')\n",
    "df = df[(df.value != \"-\") & (df.value != '.')]\n",
    "df[['variable', 'value']].groupby(['variable', 'value']).aggregate(len).unstack().plot(kind='bar')\n",
    "df.ix[df.variable != \"47\", :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 0 cases where a single base is not aligned to position 47 properly."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
