{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import subprocess\n",
    "import math\n",
    "import glob\n",
    "import os, sys\n",
    "from collections import defaultdict, Counter\n",
    "from Bio import SeqIO, SeqRecord, Seq\n",
    "from tRNA_position import *\n",
    "pd.set_option('display.max_colwidth',10000)\n",
    "pd.set_option('display.width', 10000)\n",
    "pd.set_option('display.max_columns', 10000)\n",
    "pd.set_option('display.max_rows',1000)\n",
    "isotypes = ['Ala', 'Arg', 'Asn', 'Asp', 'Cys', 'Gln', 'Glu', 'Gly', 'His', 'Ile', 'iMet', 'Leu', 'Lys', 'Met', 'Phe', 'Pro', 'Ser', 'Thr', 'Trp', 'Tyr', 'Val']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "A global view of identity elements versus biological features would be a powerful tool for predicting tRNA function using primary sequence. To do this, I'll need to align eukaryotic tRNAs to our established models and annotate positions based on universal numbering. Then, I'll extract sequence information such as position, clade, or variable arm length."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process tRNAs\n",
    "### Species information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>shortname</th>\n",
       "      <th>name</th>\n",
       "      <th>longname</th>\n",
       "      <th>domain</th>\n",
       "      <th>clade</th>\n",
       "      <th>subclade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bdist3</td>\n",
       "      <td>braDis3</td>\n",
       "      <td>Brachypodium distachyon Bd21 (JGI v3.0)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mtrun4</td>\n",
       "      <td>medTru4</td>\n",
       "      <td>Medicago truncatula (Mt4.0v1)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Graim2</td>\n",
       "      <td>gosRai2</td>\n",
       "      <td>Gossypium raimondii (Cotton v2)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ptric3</td>\n",
       "      <td>popTri3</td>\n",
       "      <td>Populus trichocarpa (JGI v3.0)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sbico3</td>\n",
       "      <td>sorBic3</td>\n",
       "      <td>Sorghum bicolor (Version 3.1)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Oarie1</td>\n",
       "      <td>oviAri3</td>\n",
       "      <td>Ovis aries (sheep) (Feb 2010)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zmays6</td>\n",
       "      <td>zeaMay6</td>\n",
       "      <td>Zea mays B73 (Version AGPv3)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Gmax2</td>\n",
       "      <td>glyMax2</td>\n",
       "      <td>Glycine max (soybean) (Wm82.a2)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Rnorv6</td>\n",
       "      <td>rn6</td>\n",
       "      <td>Rattus norvegicus (Rat Jul. 2014 RGSC 6.0/rn6)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Vpaco2</td>\n",
       "      <td>vicPac2</td>\n",
       "      <td>Vicugna pacos (Alpaca Mar. 2013 Vicugna_pacos-2.0.1/vicPac2)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Osati7</td>\n",
       "      <td>orySat7</td>\n",
       "      <td>Oryza sativa Japonica Group (Japanese rice IRGSP-1.0)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>danRer10</td>\n",
       "      <td>danRer10</td>\n",
       "      <td>Danio rerio (Zebrafish Sep. 2014 GRCz10/danRer10)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Vertebrata</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>proCap1</td>\n",
       "      <td>proCap1</td>\n",
       "      <td>Procavia capensis (Rock hyrax Jul. 2008 Broad/proCap1)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pteVam1</td>\n",
       "      <td>pteVam1</td>\n",
       "      <td>Pteropus vampyrus (Megabat Jul. 2008 Broad/pteVam1)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Mammalia</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>manEsc6</td>\n",
       "      <td>manEsc6</td>\n",
       "      <td>Manihot esculenta (Cassava v6)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Spermatophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Pfalc</td>\n",
       "      <td>plasFalc1</td>\n",
       "      <td>Plasmodium falciparum (Oct 2007)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Apicomplexa</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ppate</td>\n",
       "      <td></td>\n",
       "      <td>Physcomitrella patens (Version 1.1)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Bryophyta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Spurp</td>\n",
       "      <td></td>\n",
       "      <td>Strongylocentrotus purpuratus (Sea urchin) (Version 2.1)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Echinodermata</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Lmajo</td>\n",
       "      <td></td>\n",
       "      <td>Leishmania major (Version 5.0)</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Euglenozoa</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ashb_goss_ATCC_10895</td>\n",
       "      <td>ashbGoss_ATCC10895</td>\n",
       "      <td>Ashbya gossypii ATCC 10895</td>\n",
       "      <td>eukaryota</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               shortname                name                                                      longname     domain          clade subclade\n",
       "0                 Bdist3             braDis3                       Brachypodium distachyon Bd21 (JGI v3.0)  eukaryota   Streptophyta      NaN\n",
       "1                 Mtrun4             medTru4                                 Medicago truncatula (Mt4.0v1)  eukaryota   Streptophyta      NaN\n",
       "2                 Graim2             gosRai2                               Gossypium raimondii (Cotton v2)  eukaryota   Streptophyta      NaN\n",
       "3                 Ptric3             popTri3                                Populus trichocarpa (JGI v3.0)  eukaryota   Streptophyta      NaN\n",
       "4                 Sbico3             sorBic3                                 Sorghum bicolor (Version 3.1)  eukaryota   Streptophyta      NaN\n",
       "5                 Oarie1             oviAri3                                 Ovis aries (sheep) (Feb 2010)  eukaryota       Mammalia      NaN\n",
       "6                 Zmays6             zeaMay6                                  Zea mays B73 (Version AGPv3)  eukaryota   Streptophyta      NaN\n",
       "7                  Gmax2             glyMax2                               Glycine max (soybean) (Wm82.a2)  eukaryota   Streptophyta      NaN\n",
       "8                 Rnorv6                 rn6                Rattus norvegicus (Rat Jul. 2014 RGSC 6.0/rn6)  eukaryota       Mammalia      NaN\n",
       "9                 Vpaco2             vicPac2  Vicugna pacos (Alpaca Mar. 2013 Vicugna_pacos-2.0.1/vicPac2)  eukaryota       Mammalia      NaN\n",
       "10                Osati7             orySat7         Oryza sativa Japonica Group (Japanese rice IRGSP-1.0)  eukaryota   Streptophyta      NaN\n",
       "11              danRer10            danRer10             Danio rerio (Zebrafish Sep. 2014 GRCz10/danRer10)  eukaryota     Vertebrata      NaN\n",
       "12               proCap1             proCap1        Procavia capensis (Rock hyrax Jul. 2008 Broad/proCap1)  eukaryota       Mammalia      NaN\n",
       "13               pteVam1             pteVam1           Pteropus vampyrus (Megabat Jul. 2008 Broad/pteVam1)  eukaryota       Mammalia      NaN\n",
       "14               manEsc6             manEsc6                                Manihot esculenta (Cassava v6)  eukaryota  Spermatophyta      NaN\n",
       "15                 Pfalc           plasFalc1                              Plasmodium falciparum (Oct 2007)  eukaryota    Apicomplexa      NaN\n",
       "16                 Ppate                                               Physcomitrella patens (Version 1.1)  eukaryota      Bryophyta      NaN\n",
       "17                 Spurp                          Strongylocentrotus purpuratus (Sea urchin) (Version 2.1)  eukaryota  Echinodermata      NaN\n",
       "18                 Lmajo                                                    Leishmania major (Version 5.0)  eukaryota     Euglenozoa      NaN\n",
       "19  Ashb_goss_ATCC_10895  ashbGoss_ATCC10895                                    Ashbya gossypii ATCC 10895  eukaryota          Fungi      NaN"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_table = pd.read_table('genome_table+.txt', header=None, names=['shortname', 'name', 'longname', 'domain', 'clade', 'subclade'])\n",
    "# set NaN to empty strings\n",
    "species_table.ix[species_table.name.apply(lambda x: type(x) is float), 'name'] = ''\n",
    "species_table.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align tRNAs to existing CMs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have a \"quality set\" of tRNAs, which include tRNAs from previous versions of genomes and newer versions of genomes. It is simpler and easier to create a new fasta file and alignment for each isotype, based on the `.iso` files from each run.\n",
    "\n",
    "I also need to manually curate the genome table file to include the new species. And while I'm doing that, I might as well fix the broken entries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing Mus129S1_SvImJ_1509...Mus129S1_SvImJ_1509 tRNA file not found, skipping\n",
      "processing MusAKR_J_1509...MusAKR_J_1509 tRNA file not found, skipping\n",
      "processing MusA_J_1509...MusA_J_1509 tRNA file not found, skipping\n",
      "processing MusBALB_cJ_1509...MusBALB_cJ_1509 tRNA file not found, skipping\n",
      "processing MusC3H_HeJ_1509...MusC3H_HeJ_1509 tRNA file not found, skipping\n",
      "processing MusC57BL_6NJ_1509...MusC57BL_6NJ_1509 tRNA file not found, skipping\n",
      "processing MusCAROLI_EiJ_1509...MusCAROLI_EiJ_1509 tRNA file not found, skipping\n",
      "processing MusCAST_EiJ_1509...MusCAST_EiJ_1509 tRNA file not found, skipping\n",
      "processing MusCBA_J_1509...MusCBA_J_1509 tRNA file not found, skipping\n",
      "processing MusDBA_2J_1509...MusDBA_2J_1509 tRNA file not found, skipping\n",
      "processing MusFVB_NJ_1509...MusFVB_NJ_1509 tRNA file not found, skipping\n",
      "processing MusLP_J_1509...MusLP_J_1509 tRNA file not found, skipping\n",
      "processing MusNOD_ShiLtJ_1509...MusNOD_ShiLtJ_1509 tRNA file not found, skipping\n",
      "processing MusNZO_HlLtJ_1509...MusNZO_HlLtJ_1509 tRNA file not found, skipping\n",
      "processing MusPWK_PhJ_1509...MusPWK_PhJ_1509 tRNA file not found, skipping\n",
      "processing MusPahari_EiJ_1509...MusPahari_EiJ_1509 tRNA file not found, skipping\n",
      "processing MusSPRET_EiJ_1509...MusSPRET_EiJ_1509 tRNA file not found, skipping\n",
      "processing MusWSB_EiJ_1509...MusWSB_EiJ_1509 tRNA file not found, skipping\n",
      "processing ailMel1...finished\n",
      "processing anoCar2...finished\n",
      "processing araTha1...finished\n",
      "processing ashbGoss_ATCC10895...finished\n",
      "processing aspeFumi_AF293...finished\n",
      "processing aspeNidu_FGSC_A4...finished\n",
      "processing aspeOryz_RIB40...finished\n",
      "processing balAcu1...finished\n",
      "processing bosTau8...finished\n",
      "processing botrCine_B05_10...finished\n",
      "processing braDis3...finished\n",
      "processing caeJap1...finished\n",
      "processing caePb2...finished\n",
      "processing caeRem3...finished\n",
      "processing calJac3...finished\n",
      "processing calMil1...finished\n",
      "processing canFam3...finished\n",
      "processing candAlbi_WO_1...finished\n",
      "processing candDubl_CD36...finished\n",
      "processing candGlab_CBS_138...finished\n",
      "processing candOrth_CO_90_125...finished\n",
      "processing cavPor3...finished\n",
      "processing cb3...finished\n",
      "processing ce10...finished\n",
      "processing cerSim1...finished\n",
      "processing chaeTher_VAR_THERMOPHILUM_DSM1...chaeTher_VAR_THERMOPHILUM_DSM1 tRNA file not found, skipping\n",
      "processing chrPic1...finished\n",
      "processing criGri1...finished\n",
      "processing crypGatt_WM276...finished\n",
      "processing crypNeof_VAR_GRUBII_H99...finished\n",
      "processing crypNeof_VAR_NEOFORMANS_B_3501...finished\n",
      "processing crypNeof_VAR_NEOFORMANS_JEC21...finished\n",
      "processing cybeJadi_NBRC_0988_NBRC0988...finished\n",
      "processing danRer10...finished\n",
      "processing dasNov3...finished\n",
      "processing debaHans_CBS767...finished\n",
      "processing dipOrd1...dipOrd1 tRNA file not found, skipping\n",
      "processing dm6...finished\n",
      "processing dp4...finished\n",
      "processing droAna3...finished\n",
      "processing droEre2...finished\n",
      "processing droGri2...finished\n",
      "processing droMoj3...finished\n",
      "processing droPer1...finished\n",
      "processing droSec1...finished\n",
      "processing droSim1...finished\n",
      "processing droVir3...finished\n",
      "processing droWil1...finished\n",
      "processing droYak2...finished\n",
      "processing enceCuni_GB_M1...finished\n",
      "processing enceHell_ATCC50504...finished\n",
      "processing enceInte_ATCC50506...finished\n",
      "processing enceRoma_SJ_2008...finished\n",
      "processing equCab2...finished\n",
      "processing eremCymb_DBVPG#7215...finished\n",
      "processing eriEur2...finished\n",
      "processing felCat5...finished\n",
      "processing flamVelu_KACC42780...finished\n",
      "processing fr3...finished\n",
      "processing fusaGram_CS3005...finished\n",
      "processing fusaGram_PH_1_NRRL_31084...finished\n",
      "processing fusaPseu_CS3096...finished\n",
      "processing fusaPseu_CS3270...finished\n",
      "processing fusaVert_7600...finished\n",
      "processing gadMor1...finished\n",
      "processing galGal4...finished\n",
      "processing gasAcu1...finished\n",
      "processing geoFor1...finished\n",
      "processing glyMax2...finished\n",
      "processing gorGor3...finished\n",
      "processing gosRai2...finished\n",
      "processing hetGla2...finished\n",
      "processing hg19...finished\n",
      "processing hg38...finished\n",
      "processing kazaAfri_CBS_2517...finished\n",
      "processing kazaNaga_CBS_8797...finished\n",
      "processing kluyLact_NRRL_Y_1140...finished\n",
      "processing komaPast_CBS_7435...finished\n",
      "processing komaPast_GS115...finished\n",
      "processing lachKluy_NRRL_Y_12651...finished\n",
      "processing lachTher_CBS_6340...finished\n",
      "processing latCha1...finished\n",
      "processing loxAfr3...finished\n",
      "processing macEug2...finished\n",
      "processing magnOryz_70_15...finished\n",
      "processing manEsc6...finished\n",
      "processing medTru4...finished\n",
      "processing melGal1...finished\n",
      "processing melUnd1...finished\n",
      "processing micMur1...finished\n",
      "processing millFari_CBS_7064...finished\n",
      "processing mm10...finished\n",
      "processing monDom5...finished\n",
      "processing musFur1...finished\n",
      "processing myceTher_ATCC42464...finished\n",
      "processing myoLuc2...finished\n",
      "processing naumCast_CBS_4309...finished\n",
      "processing naumDair_CBS_421...finished\n",
      "processing neurCras_OR74A...finished\n",
      "processing nomLeu3...finished\n",
      "processing ochPri3...finished\n",
      "processing ogatPara_DL_1...finished\n",
      "processing oreNil2...finished\n",
      "processing ornAna1...finished\n",
      "processing oryCun2...finished\n",
      "processing oryLat2...finished\n",
      "processing orySat7...finished\n",
      "processing otoGar3...finished\n",
      "processing oviAri3...finished\n",
      "processing panTro4...finished\n",
      "processing papAnu2...finished\n",
      "processing peniChry_P2NIAD18...finished\n",
      "processing ponAbe2...finished\n",
      "processing popTri3...finished\n",
      "processing priPac1...finished\n",
      "processing proCap1...finished\n",
      "processing pteVam1...finished\n",
      "processing rheMac3...finished\n",
      "processing rn6...finished\n",
      "processing sacCer3...finished\n",
      "processing saccCere_AWRI796...finished\n",
      "processing saccCere_CEN_PK113_7D...finished\n",
      "processing saccCere_FOSTERSB...finished\n",
      "processing saccCere_P283...finished\n",
      "processing saccCere_P301...finished\n",
      "processing saccCere_R008...finished\n",
      "processing saccCere_R103...finished\n",
      "processing saccCere_SIGMA1278B...finished\n",
      "processing saccCere_UFMG_A_905...finished\n",
      "processing saccCere_VIN13...finished\n",
      "processing saccCere_VL3...finished\n",
      "processing saccCere_W303...finished\n",
      "processing saccCere_YJSH1...finished\n",
      "processing saccSp_ASHBYA_ACERI...finished\n",
      "processing saccSp_BOULARDII_17...finished\n",
      "processing saccSp_BOULARDII_ATCCMYA_796...finished\n",
      "processing saiBol1...finished\n",
      "processing sarHar1...finished\n",
      "processing scheStip_CBS_6054...finished\n",
      "processing schiPomb_972H...finished\n",
      "processing sorAra2...finished\n",
      "processing sorBic3...finished\n",
      "processing speTri2...finished\n",
      "processing sporReil_SRZ2...finished\n",
      "processing susScr3...finished\n",
      "processing taeGut2...finished\n",
      "processing tarSyr2...finished\n",
      "processing tetNig2...finished\n",
      "processing tetrBlat_CBS_6284...finished\n",
      "processing tetrPhaf_CBS_4417...finished\n",
      "processing thieTerr_NRRL_8126...finished\n",
      "processing toruDelb_CBS_1146...finished\n",
      "processing triMan1...finished\n",
      "processing trypBruc_BRUCEI_TREU927_927_4...trypBruc_BRUCEI_TREU927_927_4 tRNA file not found, skipping\n",
      "processing trypBruc_GAMBIENSE_DAL972_DAL...trypBruc_GAMBIENSE_DAL972_DAL tRNA file not found, skipping\n",
      "processing turTru2...finished\n",
      "processing ustiMayd_521...finished\n",
      "processing valsMali_03_8...finished\n",
      "processing vertDahl_JR2...finished\n",
      "processing vertDahl_VDLS_17...finished\n",
      "processing vicPac2...finished\n",
      "processing vitVin1...finished\n",
      "processing xenTro3...finished\n",
      "processing yarrLipo_CLIB122...finished\n",
      "processing yarrLipo_PO1F...finished\n",
      "processing yarrLipo_WSH_Z06...finished\n",
      "processing zeaMay6...finished\n",
      "processing zygoRoux_CBS_732...finished\n"
     ]
    }
   ],
   "source": [
    "species = sorted(glob.glob(\"iso/*.iso\"))\n",
    "species = [sp[4:-10] for sp in species]\n",
    "isotype_seqs = defaultdict(list)\n",
    "for sp in species:\n",
    "  sys.stdout.write('processing ' + sp + '...')\n",
    "  sys.stdout.flush()\n",
    "  tRNA_file = 'tRNAs/' + sp + '-tRNAs.fa'\n",
    "  if not os.path.exists(tRNA_file):\n",
    "    print(sp + ' tRNA file not found, skipping')\n",
    "    continue\n",
    "\n",
    "  seqs = []\n",
    "  seqids = []\n",
    "  for seq in SeqIO.parse(tRNA_file, 'fasta'):\n",
    "    if \"pseudogene\" in seq.description: continue\n",
    "    score = float(re.findall('Sc: [\\d\\.]+', seq.description)[0].split()[-1])\n",
    "    if score < 55: continue\n",
    "    seqs.append(seq)\n",
    "    seqids.append(seq.id)\n",
    "\n",
    "  iso_scores = pd.read_table('iso/' + sp + '-tRNAs.iso', header=0)\n",
    "  iso_scores = iso_scores[iso_scores.tRNAscanID.isin(seqids)]\n",
    "  iso_scores['best'] = iso_scores.ix[:,2:].idxmax(axis=1)\n",
    "  iso_scores['score'] = iso_scores.max(axis=1, numeric_only=True)\n",
    "  iso_scores.index = iso_scores.tRNAscanID.values\n",
    "  iso_scores = iso_scores[['best', 'score']]\n",
    "  \n",
    "  for seq in seqs:\n",
    "    isotype = iso_scores.ix[seq.id].best\n",
    "    if isotype == 'SeC' or 'mito' in isotype: continue\n",
    "    isoscore = iso_scores.ix[seq.id].score\n",
    "    seq.id = '{}_{} Iso: {}'.format(sp, seq.description, isoscore)\n",
    "    seq.description = ''\n",
    "    isotype_seqs[isotype].append(seq)\n",
    "\n",
    "  print('finished')\n",
    "  sys.stdout.flush()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Run this only after successfully running the above\n",
    "for isotype in isotypes:\n",
    "  fasta = 'fasta/euk-' + isotype + '.fa'\n",
    "  if os.path.exists(fasta): # Force user to manually check whether it needs to be rewritten\n",
    "    print(\"File exists: \" + fasta)\n",
    "    continue\n",
    "  fasta_handle = open(fasta, 'w')\n",
    "  SeqIO.write(isotype_seqs[isotype], fasta_handle, 'fasta')\n",
    "  fasta_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for isotype in isotypes: \n",
    "  fasta = 'fasta/euk-' + isotype + '.fa'\n",
    "  alignment = 'alignments/euk-' + isotype + '.sto'\n",
    "  model = '/projects/lowelab/users/blin/tRNAscan/models/current/TRNAinf-euk-' + isotype + '.cm'\n",
    "  subprocess.call('cmalign -g --notrunc -o {} {} {} > /dev/null'.format(alignment, model, fasta), shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse alignment positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing Ala..."
     ]
    }
   ],
   "source": [
    "def position_base(positions, seq):\n",
    "  for position_index, position in enumerate(positions):\n",
    "    if position.paired:\n",
    "      index1, index2 = position.position.split(':')\n",
    "      index1, index2 = int(index1), int(index2)\n",
    "      base_pair = \"{}:{}\".format(seq[index1 - 1], seq[index2 - 1])\n",
    "      yield position.sprinzl, base_pair\n",
    "    else:\n",
    "      index = int(position.position)\n",
    "      base = seq[index - 1]\n",
    "      yield position.sprinzl, base\n",
    "      \n",
    "identities = pd.DataFrame()\n",
    "\n",
    "for isotype in isotypes:\n",
    "  sys.stdout.write('processing ' + isotype + '...')\n",
    "  alignment = 'alignments/euk-' + isotype + '.sto'\n",
    "  # get positions\n",
    "  alignment_fhandle = open(alignment)\n",
    "  positions = [] # list containing each position in the tRNA\n",
    "  # first, get secondary structure\n",
    "  for line in alignment_fhandle:\n",
    "    if line[0:12] == '#=GC SS_cons':\n",
    "      ss = line.strip().split()[-1]\n",
    "  # parse secondary structure into regions and positions\n",
    "  positions = annotate_positions(ss)\n",
    "  # get counts for each position by parsing Stockholm file\n",
    "  alignment_fhandle = open(alignment)\n",
    "  for line in alignment_fhandle:\n",
    "    if line[0] in [\"#\", '\\n', '/']: continue\n",
    "    seqname, seq = line.strip().split()\n",
    "    species = re.split('_|\\\\.trna', seqname, 1)[0]\n",
    "    if any(species_table.name == species): row = species_table[species_table.name == species]\n",
    "    else: continue # skip species that need too much babying\n",
    "    domain = row.domain.values[0]\n",
    "    clade = row.clade.values[0]\n",
    "    species_long = row.longname.values[0]\n",
    "    trna = {'domain': domain, 'clade': clade, 'species': species, 'species_long': species_long, 'seqname': seqname, 'isotype': isotype}\n",
    "    trna = {**trna, **{sprinzl: base for sprinzl, base in position_base(positions, seq)}}\n",
    "    identities = identities.append([trna], ignore_index=True)\n",
    "  \n",
    "  print('done')\n",
    "  \n",
    "identities.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional sequence information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Isotype, anticodon, score\n",
    "seqinfo = []\n",
    "for isotype in isotypes:\n",
    "  alignment = 'alignments/euk-' + isotype + '.sto'\n",
    "  for line in open(alignment):\n",
    "    if line[0:4] != \"#=GS\": continue\n",
    "    _, seqname, _, _, _, isotype, anticodon, _, _, _, score, _, isoscore = line.strip().split()\n",
    "    seqinfo.append([seqname, isotype, anticodon[1:-1], float(score), float(isoscore)])\n",
    "seqinfo = pd.DataFrame(seqinfo, columns=['seqname', 'isotype_ac', 'anticodon', 'score', 'isoscore'])\n",
    "identities = identities.merge(seqinfo, on='seqname')\n",
    "\n",
    "# GC content\n",
    "paired_cols = identities.columns[list(map(lambda x: (':' in x), identities.columns))]\n",
    "identities['GC'] = identities[paired_cols].apply(lambda x: sum((x == \"G:C\") | (x == \"C:G\"))/len(paired_cols), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loop sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "def bounds_to_cols(cols, start, end):\n",
    "  selected_cols = []\n",
    "  for col in cols:\n",
    "    matches = re.findall('\\d+', col)\n",
    "    if len(matches) < 1: continue\n",
    "    index = int(matches[0])\n",
    "    if (index >= start and index <= end or col[0:3] == '{}i'.format(start - 1)) and col[0] != 'V':\n",
    "      selected_cols.append(col)\n",
    "  return selected_cols\n",
    "\n",
    "dloop_cols = bounds_to_cols(identities.columns, 14, 21)\n",
    "identities['D-loop'] = identities[dloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "acloop_cols = bounds_to_cols(identities.columns, 32, 38)\n",
    "identities['AC-loop'] = identities[acloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "tpcloop_cols = bounds_to_cols(identities.columns, 54, 60)\n",
    "identities['TPC-loop'] = identities[tpcloop_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)\n",
    "\n",
    "varm_cols = list(filter(lambda x: 'V' in x, identities.columns))\n",
    "identities['V-arm'] = identities[varm_cols].apply(lambda x: len(x[(x != '.') & (x != '-')]), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Insertions/deletions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Insertions\n",
    "insertion_cols = list(filter(lambda x: re.match('\\d+i', x), identities.columns))\n",
    "identities['insertions'] = identities[insertion_cols].apply(lambda x: sum(x != '.'), axis=1)\n",
    "\n",
    "# Deletions at positions that are not the variable arm\n",
    "base_cols = list(filter(lambda x: re.match('^((\\d+)|(\\d+:\\d+))$', x), identities.columns))\n",
    "identities['deletions'] = identities[base_cols].apply(lambda x: ''.join(x).count('-'), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### High-quality tRNA set\n",
    "\n",
    "There was talk about using the high quality tRNAs to look at identity elements. Instead, I will just annotate specific tRNAs as high-quality or not. We can sort out the differences later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Import list of quality tRNAs\n",
    "quality_trnas = [line.strip() for line in open('quality-set.out')]\n",
    "identities['quality'] = identities.seqname.isin(quality_trnas)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Restrict tRNAs by species\n",
    "\n",
    "We may also want to limit the contribution of any single species, similar to how we built the isotype-specific models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "identities.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
